---
title: "Lin41 Analysis"
author: "Micah Gearhart"
date: "1/6/2017"
output:
  html_document:
    keep_md: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r libraries,eval=T,results='hide',message=FALSE,warning=FALSE}
library("biomaRt")
library("GenomicFeatures")
library("Rsamtools")

library("GenomicFeatures")
library("Rsamtools")
library("GenomicAlignments")
library("BiocParallel")
library("DESeq2")
library("genefilter")
library("pheatmap")
library("SRAdb")
library("goseq")


library("ggplot2")
library("dplyr")
library("tidyr")
library("magrittr")
library("RColorBrewer")
library("gridExtra")
library("motifStack")

(ts<-format(Sys.time(), "%a_%b_%d_%Y_%H%M"))
cbPalette <- c("#999999", "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7")
wls<- function(x) { paste0(seqnames(x),":",start(x),"-",end(x))}
```

##Download Annotation files for Ensembl 82 and 87
```{r biomaRt, eval=F}
ensembl_82 = useMart(biomart="ENSEMBL_MART_ENSEMBL", host="sep2015.archive.ensembl.org", path="/biomart/martservice",dataset="celegans_gene_ensembl")
txdb<-makeTxDbFromBiomart(biomart="ENSEMBL_MART_ENSEMBL",host="sep2015.archive.ensembl.org",dataset="celegans_gene_ensembl")
ens82<-exonsBy(txdb,by="gene")
type82<-getBM(filters="ensembl_gene_id",values=names(ens82), attributes=c("ensembl_gene_id","transcript_biotype"), mart=ensembl_82)
table(type82$transcript_biotype)
save(ens82,type82,file="ens82.rdata")

ensembl_87 = useMart(biomart="ENSEMBL_MART_ENSEMBL", host="useast.ensembl.org", path="/biomart/martservice",dataset="celegans_gene_ensembl")
txdb<-makeTxDbFromBiomart(biomart="ENSEMBL_MART_ENSEMBL",host="useast.ensembl.org",dataset="celegans_gene_ensembl")

tx87<-transcriptsBy(txdb, "gene")
fiveUTR<-fiveUTRsByTranscript(txdb,use.names=T)
threeUTR<-threeUTRsByTranscript(txdb,use.names=T)

ens87<-exonsBy(txdb,by="gene")
type87<-getBM(filters="ensembl_gene_id",values=names(ens87), attributes=c("ensembl_gene_id","transcript_biotype"), mart=ensembl_87)
table(type87$transcript_biotype)
save(ens87,type87,file="ens87.rdata")
```

```{r}
countBam("./2566IP-1_S21.bam",param=ScanBamParam(flag = scanBamFlag(isUnmappedQuery = TRUE),))

mapq_filter <- function(features, reads, ignore.strand = FALSE, inter.feature=TRUE)
{
    require(GenomicAlignments) # needed for parallel evaluation
    Union(features, reads[mcols(reads)$mapq >= 4],
          ignore.strand, inter.feature)
}

param <- ScanBamParam(what="mapq")

#select single end bam files that are not filtered by samtools (q4)
(fls <- list.files(".", pattern=glob2rx("*.bam"),full=TRUE))
(fls<-fls[!grepl("q4",fls)])
(fls<-fls[!grepl("PE",fls)])
bamlst <- BamFileList(fls)

#summarizeOverlaps with multimappers
genehits82<-summarizeOverlaps(ens82,bamlst,mode="Union",singleEnd=TRUE,ignore.strand=TRUE)
apply(assays(genehits82)$counts,2,sum)

#with multimapping Filter
genehits82_mapq_filter<-summarizeOverlaps(ens82,bamlst,mode="mapq_filter", param=param,
                                 singleEnd=TRUE,ignore.strand=TRUE)
apply(assays(genehits82_mapq_filter)$counts,2,sum)

#just check to make sure this is the same as samtools filtering
(fls <- list.files(".", pattern=glob2rx("*_q4.bam"),full=TRUE))
bamlst <- BamFileList(fls)
genehits82q4<-summarizeOverlaps(ens82,bamlst,mode="Union",singleEnd=TRUE,ignore.strand=TRUE)
apply(assays(genehits82q4)$counts,2,sum)

all.equal(unname(apply(assays(genehits82_mapq_filter)$counts,2,sum)),unname(apply(assays(genehits82q4)$counts,2,sum)))

save(genehits82,genehits82_mapq_filter,file="genehits82_Mon_Jan_30_2017_2210.rdata")

```


#Count 2x for Ensembl 87
```{r}
load("ens87.rdata")
mapq_filter <- function(features, reads, ignore.strand = FALSE, inter.feature=TRUE)
{
    require(GenomicAlignments) # needed for parallel evaluation
    Union(features, reads[mcols(reads)$mapq >= 4],
          ignore.strand, inter.feature)
}

param <- ScanBamParam(what="mapq")

#select single end bam files that are not filtered by samtools (q4)
(fls <- list.files(".", pattern=glob2rx("*.bam"),full=TRUE))
(fls<-fls[!grepl("q4",fls)])
(fls<-fls[!grepl("PE",fls)])
(fls<-fls[!grepl("IP4_4_23_10.bam",fls)])
bamlst <- BamFileList(fls)

#summarizeOverlaps with multimappers
genehits87<-summarizeOverlaps(ens87,bamlst,mode="Union",singleEnd=TRUE,ignore.strand=TRUE)
apply(assays(genehits87)$counts,2,sum)

#with multimapping Filter
genehits87_mapq_filter<-summarizeOverlaps(ens87,bamlst,mode="mapq_filter", param=param,
                                 singleEnd=TRUE,ignore.strand=TRUE)
apply(assays(genehits87_mapq_filter)$counts,2,sum)

fiveUTRhits87_mapq_filter<-summarizeOverlaps(fiveUTR,bamlst,mode="mapq_filter", param=param,
                                 singleEnd=TRUE,ignore.strand=TRUE)
apply(assays(fiveUTRhits87_mapq_filter)$counts,2,sum)
threeUTRhits87_mapq_filter<-summarizeOverlaps(threeUTR,bamlst,mode="mapq_filter", param=param,
                                 singleEnd=TRUE,ignore.strand=TRUE)
apply(assays(threeUTRhits87_mapq_filter)$counts,2,sum)

#just check to make sure this is the same as samtools filtering
#(fls <- list.files(".", pattern=glob2rx("*_q4.bam"),full=TRUE))
#bamlst <- BamFileList(fls)
#genehits82q4<-summarizeOverlaps(ens82,bamlst,mode="Union",singleEnd=TRUE,ignore.strand=TRUE)
#apply(assays(genehits82q4)$counts,2,sum)
#all.equal(unname(apply(assays(genehits82_mapq_filter)$counts,2,sum)),unname(apply(assays(genehits82q4)$counts,2,sum)))

save(genehits87,genehits87_mapq_filter,file=paste0("genehits87_",ts,".rdata")
```

## Compare the biotype Distributions across samples with and without a MAPQ filter
```{r}
load("genehits87_Tue_Jan_30_2017_1610.rdata")

counts87<-as.data.frame(assays(genehits87)$counts)
counts87q4<-as.data.frame(assays(genehits87_mapq_filter)$counts)
                          
summary(idx<-match(rownames(counts87),type87$ensembl_gene_id))
counts87$biotype<-type87[idx,"transcript_biotype"]
counts87q4$biotype<-type87[idx,"transcript_biotype"]

g1<-gather(counts87,sample,counts,-biotype) %>%
  group_by(sample,biotype) %>%
  summarize(countSum=sum(counts)) %>%
  ggplot(aes(x=sample,y=countSum,fill=biotype)) + ggtitle("Counts with Ensembl87") +
  scale_fill_manual(values=rev(colors),guide = guide_legend(reverse=TRUE)) +
  geom_bar(stat="identity") + theme_classic() + theme(axis.text.x = element_text(angle = 90,  hjust = 0)) 

g2<-gather(counts87q4,sample,counts,-biotype) %>%
  group_by(sample,biotype) %>%
  summarize(countSum=sum(counts)) %>%
  ggplot(aes(x=sample,y=countSum,fill=biotype)) + ggtitle("Counts with Ensembl87 with Mapping Quality Filter") +
  scale_fill_manual(values=rev(colors),guide = guide_legend(reverse=TRUE)) +
  geom_bar(stat="identity") + theme_classic() + theme(axis.text.x = element_text(angle = 90,  hjust = 0)) 

grid.arrange(g1,g2,ncol=1)
```


# Filter out rRNA genes and analyze for Enrichment relative to Input
```{r}
summary(idx<-match(rownames(genehits87),type87$ensembl_gene_id))
dim(dds87<-DESeqDataSet(genehits87[type87[idx,]$transcript_biotype != "rRNA",],design=~1))
dim(dds87q4<-DESeqDataSet(genehits87_mapq_filter[type87[idx,]$transcript_biotype != "rRNA",],design=~1))

file_names<-colnames(dds87)
colnames(dds87)<-sapply(strsplit(colnames(dds87),"_"),function(x) x[1]) %>%
  gsub("2566","OMA-1_",.) %>% gsub("3923","LIN-41_",.) %>% gsub("IP4","OMA-1_042310",.)

colnames(dds87q4)<-sapply(strsplit(colnames(dds87q4),"_"),function(x) x[1]) %>%
  gsub("2566","OMA-1_",.) %>% gsub("3923","LIN-41_",.) %>% gsub("IP4","OMA-1_042310",.)

dds87$tag<-as.factor(c("OMA-1","OMA-1","LysRZ","Lys","LIN-41","LIN-41","LysRZ","Lys","OMA-1_042310"))
dds87q4$tag<-as.factor(c("OMA-1","OMA-1","LysRZ","Lys","LIN-41","LIN-41","LysRZ","Lys","OMA-1_042310"))

design(dds87)<-(~tag)
design(dds87q4)<-(~tag)

plotPCA( DESeqTransform( dds87 ),intgroup="tag")+ ggtitle("PCA without MapQ Filter") + theme_bw()
plotPCA( DESeqTransform( dds87q4 ),intgroup="tag")+ ggtitle("PCA with MapQ Filter") + theme_bw()

```

## Run Differential Expression on both dds objects
```{r}
dds87<-DESeq(dds87)
plotMA(dds87,ylim=c(-8,8))
resultsNames(dds87)

res_Oma1_dds87<-results(dds87,lfcThreshold=2,contrast=c("tag","OMA-1","LysRZ"))
res_Lin41_dds87<-results(dds87,lfcThreshold=2,contrast=c("tag","LIN-41","LysRZ"))

#Estimate LogFC for single replicate Oma-1 sample
dds87_counts<-counts(dds87,normalized=TRUE)
res_oma1_042310_dds87<-log2(dds87_counts[,9]+1)-0.5*log2(dds87_counts[,3]+1)-0.5*log2(dds87_counts[,7]+1)

res_dds87<-cbind(res_oma1_042310_dds87,res_Oma1_dds87$log2FoldChange,res_Lin41_dds87$log2FoldChange)
colnames(res_dds87)<-c("Oma-1 042310","Oma-1 IP","Lin-41 IP")

dds87q4<-DESeq(dds87q4)
plotMA(dds87q4,ylim=c(-8,8))
resultsNames(dds87q4)

res_Oma1_dds87q4<-results(dds87q4,lfcThreshold=2,contrast=c("tag","OMA-1","LysRZ"))
res_Lin41_dds87q4<-results(dds87q4,lfcThreshold=2,contrast=c("tag","LIN-41","LysRZ"))

#Estimate LogFC for single replicate Oma-1 sample
dds87q4_counts<-counts(dds87q4,normalized=TRUE)
res_oma1_042310_dds87q4<-log2(dds87q4_counts[,9]+1)-0.5*log2(dds87q4_counts[,3]+1)-0.5*log2(dds87q4_counts[,7]+1)

res_dds87q4<-cbind(res_oma1_042310_dds87q4,res_Oma1_dds87q4$log2FoldChange,res_Lin41_dds87q4$log2FoldChange)
colnames(res_dds87q4)<-c("Oma-1 042310","Oma-1 IP","Lin-41 IP")
```


#Venn Diagrams - with p-values
```{r}

lfc<-2  # also note not using abs() here -- just want enrichment
p<-0.05

x1<-length(oma1_genes<-rownames(subset(res_Oma1_dds87,log2FoldChange > lfc & padj < p)))
y1<-length(lin41_genes<-rownames(subset(res_Lin41_dds87,log2FoldChange > lfc & padj < p)))
z1<-sum(oma1_genes %in% lin41_genes)
grid.newpage()
vennplot <- draw.pairwise.venn(x1,y1,z1, c("Oma-1", "Lin-41"))

x2<-length(oma1_genesq4<-rownames(subset(res_Oma1_dds87q4,log2FoldChange > lfc & padj < p)))
y2<-length(lin41_genesq4<-rownames(subset(res_Lin41_dds87q4,log2FoldChange > lfc & padj < p)))
z2<-length(overlapq4<-oma1_genesq4[oma1_genesq4 %in% lin41_genesq4])
x21<-length(oma1_genesq4_exclusive<-oma1_genesq4[!oma1_genesq4 %in% overlapq4])
y21<-length(lin41_genesq4_exclusive<-lin41_genesq4[!lin41_genesq4 %in% overlapq4])
grid.newpage()
vennplot <- draw.pairwise.venn(x2,y2,z2, c("Oma-1", "Lin-41"))
```

#


```{r}
lfc<-2
summary(res_dds87)

#remove the NAs
temp<-as.data.frame(res_dds87[!(is.na(res_dds87[,2]) | is.na(res_dds87[,3])),])
tempq4<-as.data.frame(res_dds87q4[!(is.na(res_dds87q4[,2]) | is.na(res_dds87q4[,3])),])

(x1<-nrow(temp[temp$`Lin-41 IP`> lfc,]))
(y1<-nrow(temp[temp$`Oma-1 IP` > lfc,]))
(z1<-nrow(temp[temp$`Lin-41 IP` > lfc & temp$`Oma-1 IP` > lfc,]))

(x2<-nrow(tempq4[tempq4$`Lin-41 IP`> lfc,]))
(y2<-nrow(tempq4[tempq4$`Oma-1 IP` > lfc,]))
(z2<-nrow(tempq4[tempq4$`Lin-41 IP` > lfc & tempq4$`Oma-1 IP` > lfc,]))

#svglite("testplot.svg")
grid.newpage()
vennplot <- draw.pairwise.venn(x1,y1,z1, c("Lin-41", "Oma-1"))
grid.newpage()
vennplot <- draw.pairwise.venn(x2,y2,z2, c("Lin-41", "Oma-1"))
#dev.off()
```

#Now run GO Analysis on each section of Venn Diagram
```{r}
load("wb_257_goterms.rdata")
length(expressed_genes<-rownames(res_Lin41_dds87q4[res_Lin41_dds87q4$baseMean > 1,]))
#length(expressed_genes<-rownames(res_Lin41_dds87q4[res_Lin41_dds87q4$baseMean > 1 & !is.na(res_Lin41_dds87q4$log2FoldChange) & !is.na(res_Lin41_dds87q4$padj) & !is.na(res_Oma1_dds87q4$log2FoldChange) & !is.na(res_Oma1_dds87q4$padj),]))

dim(wb_go<-wb_go[wb_go$ensembl %in% expressed_genes,])
wb_go.list<-split(wb_go$`GO ID`,wb_go$ensembl)


#bias data
bd<-sum(width(reduce(ens87)))
bd<-bd[names(bd) %in% expressed_genes]
length(bd)
length(expressed_genes)
head(bd)

#degs
table(degs<-as.numeric(expressed_genes %in% lin41_genesq4_exclusive ))
names(degs)<-expressed_genes

table(degs<-as.numeric(expressed_genes %in% oma1_genesq4_exclusive ))
names(degs)<-expressed_genes

table(degs<-as.numeric(expressed_genes %in% overlapq4 ))
names(degs)<-expressed_genes

listGO<-function(goid) {
  print(go_terms[goid])
  tg<-wb_go[grep(goid,wb_go$`GO ID`),"ensembl",drop=F]
  idx<-match(tg$ensembl,annotations$ensembl_gene_id)
  tg$symbol<-annotations[idx,"wikigene_name"]
  tg$deg<-degs[tg$ensembl]
  as.data.frame(tg)
}

listGO("GO:0045138")
listGO("GO:1903864")
listGO("GO:0007219")

#GOseq
pwf<-nullp(degs,bias.data=bd)
GO.wall<-goseq(pwf,gene2cat=wb_go.list)
GO.wall[1:20,c("category","term","numDEInCat","numInCat","over_represented_pvalue")]

head(GO.wall,20) %>%
  dplyr::mutate(term=factor(term,levels=rev(term))) %>%
ggplot(aes(x=term,y=-log10(over_represented_pvalue))) +
  geom_bar(stat="identity",fill="red") +
  coord_flip() + xlab("") +
  theme_bw() 
```

# Download UTR sequences from biomaRt
```{r utr_biomart,eval=F}
#get transcript identifiers for all expressed genes
expressed_transcripts<-getBM(filters="ensembl_gene_id",values=expressed_genes, attributes=c("ensembl_gene_id","ensembl_transcript_id"), mart=ensembl_87)

seq3utr<-getSequence(id = expressed_transcripts$ensembl_transcript_id, type = "ensembl_transcript_id", seqType = "3utr", mart = ensembl_87)
seq5utr<-getSequence(id = expressed_transcripts$ensembl_transcript_id, type = "ensembl_transcript_id", seqType = "5utr", mart = ensembl_87)
save(seq3utr,seq5utr,file="utr_sequences.rdata")
```



# Search for OMA-1 and LIN-41 Motifs in 3' UTRs
```{r threeUTR}
load("utr_sequences.rdata")

#clean up UTR sequences
seq3utr<-seq3utr[seq3utr$`3utr`!="Sequence unavailable",]
seq3utr<-seq3utr[nchar(seq3utr$`3utr`) >=8,]  # get rid of really short 3'UTRs

# add geneid to dataframe
summary(idx<-match(seq3utr$ensembl_transcript_id,expressed_transcripts$ensembl_transcript_id))
seq3utr$ensembl_gene_id<-expressed_transcripts[idx,"ensembl_gene_id"]
sum(duplicated(paste0(seq3utr$`3utr`,seq3utr$ensembl_transcript_id)))
sum(duplicated(paste0(seq3utr$`3utr`,seq3utr$ensembl_gene_id)))
seq3utr<-seq3utr[!duplicated(paste0(seq3utr$`3utr`,seq3utr$ensembl_gene_id)),]

seq3utrBS<-DNAStringSet(seq3utr$`3utr`)
names(seq3utrBS)<-seq3utr$ensembl_transcript_id
summary(width(seq3utrBS))
seq3utr<-seq3utr[,-1]  #drop sequence from dataframe

#Count TAW motif and RNACompete Motif in each transcript
seq3utr$taw<-vcountPattern("TAW",seq3utrBS,fixed=FALSE)
table(as.logical(seq3utr$taw))

#Count RNAcompete motif in each transcript
rnaCompeteMotif<-RNAStringSet(c("GCAAAGC","GUAAAAC","GUAUAAC","GCAUAGC","CUUUAAG"))
temp<-consensusMatrix(rnaCompeteMotif)[1:4,]
plotMotifLogo(new("pfm",mat=t(t(temp[1:4,])*1/colSums(temp[1:4,])), name="RNAcompeteMotif"))

lin41_consensusMatrix<-consensusMatrix(DNAStringSet(rnaCompeteMotif))[1:4,]

seq3utr$loedige<-unlist(lapply(seq3utrBS,function(x) countPWM(lin41_consensusMatrix,x,min.score="90%")))
table(as.logical(seq3utr$loedige))

#oma-1 TAW
temp<-seq3utr %>% group_by(ensembl_gene_id) %>% 
  summarize(taw_log=as.logical(sum(taw)),loedige_log=as.logical(sum(loedige))) %>% as.data.frame

a<-sum(temp[temp$ensembl_gene_id %in% oma1_genes,"taw_log"])
b<-sum(temp[!temp$ensembl_gene_id %in% oma1_genes,"taw_log"])
c<-sum(oma1_genes %in% temp$ensembl_gene_id)
d<-sum(expressed_genes %in% temp$ensembl_gene_id)

(x<-matrix(c(d-b-c,b,c-a,a),ncol=2))
(oma1_taw_3p_test<-fisher.test(x))

#oma-1 loedige
a<-sum(temp[temp$ensembl_gene_id %in% oma1_genes,"loedige_log"])
b<-sum(temp[!temp$ensembl_gene_id %in% oma1_genes,"loedige_log"])
c<-sum(oma1_genes %in% temp$ensembl_gene_id)
d<-sum(expressed_genes %in% temp$ensembl_gene_id)

(x<-matrix(c(d-b-c,b,c-a,a),ncol=2))
(oma1_loedige_3p_test<-fisher.test(x))

#lin41 TAW
a<-sum(temp[temp$ensembl_gene_id %in% lin41_genes,"taw_log"])
b<-sum(temp[!temp$ensembl_gene_id %in% lin41_genes,"taw_log"])
c<-sum(lin41_genes %in% temp$ensembl_gene_id)
d<-sum(expressed_genes %in% temp$ensembl_gene_id)

(x<-matrix(c(d-b-c,b,c-a,a),ncol=2))
(lin41_taw_3p_test<-fisher.test(x))

#lin41 loedige
a<-sum(temp[temp$ensembl_gene_id %in% lin41_genes,"loedige_log"])
b<-sum(temp[!temp$ensembl_gene_id %in% lin41_genes,"loedige_log"])
c<-sum(lin41_genes %in% temp$ensembl_gene_id)
d<-sum(expressed_genes %in% temp$ensembl_gene_id)

(x<-matrix(c(d-b-c,b,c-a,a),ncol=2))
(lin41_loedige_3p_test<-fisher.test(x))

print(paste0("TAW motif enrichment in 3' UTR of Oma1 Enriched Genes:  ",oma1_taw_test$p.value))
print(paste0("RNAcompete Lin-41 motif enrichment in 3' UTR of Oma1 Enriched Genes:  ",oma1_loedige_test$p.value))
print(paste0("TAW motif enrichment in 3' UTR of Lin-41 Enriched Genes:  ",lin41_taw_test$p.value))
print(paste0("RNACompete Lin-41 motif enrichment in 3' UTR of Lin-41 Enriched Genes:  ",lin41_loedige_test$p.value))
```

# # Search for OMA-1 and LIN-41 Motifs in 5' UTRs 
```{r fiveUTR}
#clean up UTR sequences
seq5utr<-seq5utr[seq5utr$`5utr`!="Sequence unavailable",]
seq5utr<-seq5utr[nchar(seq5utr$`5utr`) >=8,]  # get rid of really short 3'UTRs

# add geneid to dataframe
summary(idx<-match(seq5utr$ensembl_transcript_id,expressed_transcripts$ensembl_transcript_id))
seq5utr$ensembl_gene_id<-expressed_transcripts[idx,"ensembl_gene_id"]
sum(duplicated(paste0(seq5utr$`5utr`,seq5utr$ensembl_transcript_id)))
sum(duplicated(paste0(seq5utr$`5utr`,seq5utr$ensembl_gene_id)))
seq5utr<-seq5utr[!duplicated(paste0(seq5utr$`5utr`,seq5utr$ensembl_gene_id)),]

seq5utrBS<-DNAStringSet(seq5utr$`5utr`)
names(seq5utrBS)<-seq5utr$ensembl_transcript_id
summary(width(seq5utrBS))
seq5utr<-seq5utr[,-1]  #drop sequence from dataframe

#Count TAW motif and RNACompete Motif in each transcript
seq5utr$taw<-vcountPattern("TAW",seq5utrBS,fixed=FALSE)
table(as.logical(seq5utr$taw))

#Count RNAcompete motif in each transcript
seq5utr$loedige<-unlist(lapply(seq5utrBS,function(x) countPWM(lin41_consensusMatrix,x,min.score="90%")))
table(as.logical(seq5utr$loedige))

#oma-1 TAW
temp<-seq5utr %>% group_by(ensembl_gene_id) %>% 
  summarize(taw_log=as.logical(sum(taw)),loedige_log=as.logical(sum(loedige))) %>% as.data.frame

a<-sum(temp[temp$ensembl_gene_id %in% oma1_genes,"taw_log"])
b<-sum(temp[!temp$ensembl_gene_id %in% oma1_genes,"taw_log"])
c<-sum(oma1_genes %in% temp$ensembl_gene_id)
d<-sum(expressed_genes %in% temp$ensembl_gene_id)

(x<-matrix(c(d-b-c,b,c-a,a),ncol=2))
(oma1_taw_5p_test<-fisher.test(x))

#oma-1 loedige
a<-sum(temp[temp$ensembl_gene_id %in% oma1_genes,"loedige_log"])
b<-sum(temp[!temp$ensembl_gene_id %in% oma1_genes,"loedige_log"])
c<-sum(oma1_genes %in% temp$ensembl_gene_id)
d<-sum(expressed_genes %in% temp$ensembl_gene_id)

(x<-matrix(c(d-b-c,b,c-a,a),ncol=2))
(oma1_loedige_5p_test<-fisher.test(x))

#lin41 TAW
a<-sum(temp[temp$ensembl_gene_id %in% lin41_genes,"taw_log"])
b<-sum(temp[!temp$ensembl_gene_id %in% lin41_genes,"taw_log"])
c<-sum(lin41_genes %in% temp$ensembl_gene_id)
d<-sum(expressed_genes %in% temp$ensembl_gene_id)

(x<-matrix(c(d-b-c,b,c-a,a),ncol=2))
(lin41_taw_5p_test<-fisher.test(x))

#lin41 loedige
a<-sum(temp[temp$ensembl_gene_id %in% lin41_genes,"loedige_log"])
b<-sum(temp[!temp$ensembl_gene_id %in% lin41_genes,"loedige_log"])
c<-sum(lin41_genes %in% temp$ensembl_gene_id)
d<-sum(expressed_genes %in% temp$ensembl_gene_id)

(x<-matrix(c(d-b-c,b,c-a,a),ncol=2))
(lin41_loedige_5p_test<-fisher.test(x))

print(paste0("TAW motif enrichment in 5' UTR of Oma1 Enriched Genes:  ",oma1_taw_5p_test$p.value))
print(paste0("RNAcompete Lin-41 motif enrichment in 5' UTR of Oma1 Enriched Genes:  ",oma1_loedige_5p_test$p.value))
print(paste0("TAW motif enrichment in Lin-41 5' UTR of Enriched Genes:  ",lin41_taw_5p_test$p.value))
print(paste0("RNACompete Lin-41 motif enrichment in 5' UTR of Lin-41 Enriched Genes:  ",lin41_loedige_5p_test$p.value))
```



# explore UTR's of LIN-41 exclusive set
```{r}

#count UA[A/U]  "TAW" motif in each 5'utr
seq5utr<-getSequence(id = lin41_genesq4_exclusive, type = "ensembl_gene_id", seqType = "5utr", mart = ensembl_87)
seq5utr<-seq5utr[seq5utr$`5utr`!="Sequence unavailable",]
summary(nchar(seq5utr$`5utr`))
seq5utrBS<-DNAStringSet(seq5utr$`5utr`)
seq5utr$taw<-vcountPattern("TAW",seq5utrBS,fixed=FALSE)

#count UA[A/U]  "TAW" motif in each 5'utr
seq3utr<-getSequence(id = lin41_genesq4_exclusive, type = "ensembl_gene_id", seqType = "3utr", mart = ensembl_87)
seq3utr<-seq3utr[seq3utr$`3utr`!="Sequence unavailable",]
summary(nchar(seq3utr$`3utr`))

#count UA[A/U]  "TAW" motif in each 5'utr
seq3utrBS<-DNAStringSet(seq3utr$`3utr`)
seq3utr$taw<-vcountPattern("TAW",seq3utrBS,fixed=FALSE)

rnaCompeteMotif<-RNAStringSet(c("GCAAAGC","GUAAAAC","GUAUAAC","GCAUAGC","CUUUAAG"))
temp<-consensusMatrix(rnaCompeteMotif)[1:4,]
plotMotifLogo(new("pfm",mat=t(t(temp[1:4,])*1/colSums(temp[1:4,])), name="RNAcompeteMotif"))

temp<-consensusMatrix(DNAStringSet(rnaCompeteMotif))[1:4,]
x<-countPWM(temp,seq3utrBS[[1]],min.score="80%", with.score=FALSE)

table(x<-unlist(lapply(seq3utrBS,function(x) countPWM(temp,x,min.score="80%"))))
table(y<-unlist(lapply(seq5utrBS,function(x) countPWM(temp,x,min.score="80%"))))
```

#meme
```{r}
motifSet <- runMEME(file.path(system.file("extdata",package="TFBSTools"), "crp0.s"),
                    binary="/usr/ngs/bin/meme/bin/meme",arguments=list("-nmotifs"=3))
x<-readDNAStringSet(file.path(system.file("extdata",package="TFBSTools"), "crp0.s"))

motifSet <- runMEME(x, binary="/usr/ngs/bin/meme/bin/meme",arguments=list("-nmotifs"=3))

names(seq3utrBS)<-paste0("seq3utr_",1:length(seq3utrBS))
summary(width(seq3utrBS))
seq3utrBS<-seq3utrBS[width(seq3utrBS)>= 8]
writeXStringSet(seq3utrBS,file="test3p.fasta")
motifSet_3utr <- runMEME(seq3utrBS, binary="/usr/ngs/bin/meme/bin/meme",arguments=list("-nmotifs"=3,"-maxsize"=10000000000000000,"-w"=5))
motifSet_3utr <- runMEME(seq3utrBS, binary="/usr/ngs/bin/meme/bin/meme",arguments=list("-nmotifs"=3,"-maxsize"=10000000000000000))

(temp<-consensusMatrix(motifSet_3utr)[[2]])
rownames(temp)<-c("A","C","G","U")
plotMotifLogo(new("pfm",mat=t(t(temp[1:4,])*1/colSums(temp[1:4,])), name="3'UTR motif"))

seq5utrBS<-seq5utrBS[width(seq5utrBS)>= 8]
names(seq5utrBS)<-paste0("seq5utr_",1:length(seq5utrBS))
writeXStringSet(seq5utrBS,file="test5p.fasta")
motifSet_5utr <- runMEME(seq5utrBS, binary="/usr/ngs/bin/meme/bin/meme",arguments=list("-nmotifs"=3,"-maxsize"=10000000000000000,"-w"=5))
motifSet_5utr <- runMEME(seq5utrBS, binary="/usr/ngs/bin/meme/bin/meme",arguments=list("-nmotifs"=3,"-maxsize"=10000000000000000))
consensusMatrix(motifSet_5utr)

(temp<-consensusMatrix(motifSet_5utr)[[3]])
rownames(temp)<-c("A","C","G","U")
plotMotifLogo(new("pfm",mat=t(t(temp[1:4,])*1/colSums(temp[1:4,])), name="3'UTR motif"))

```


## ###########  #########

#Is there a clever way to determine which transcript is expressed to know which UTR to use?
```{r}
d<-res_dds87q4[,]
plot(d[,3],d[,2],cex=0.5,pch=16,ylim=c(-7,7),xlim=c(-7,7),
     col=ifelse(rownames(d) %in% overlapq4,"blue",
                ifelse(rownames(d) %in% oma1_genesq4_exclusive,"green",
                ifelse(rownames(d) %in% lin41_genesq4_exclusive,"red","black"))))
abline(a=0,b=1,col="red")

d<-as.data.frame(res_dds87q4[lin41_genesq4_exclusive,])
idx<-match(rownames(d),annotations$ensembl_gene_id)
d$wiki<-annotations[idx,"wikigene_name"]
d$egn<-annotations[idx,"external_gene_name"]
d$biotype<-annotations[idx,"transcript_biotype"]
d<-d[with(d,order(-`Lin-41 IP`)),]
View(d)

getSequence(rownames(d)[1:10], seqType = "5utr", mart = ensembl_87)

tx87<-transcriptsBy(txdb, "gene")
unlist(tx87[rownames(d[1:6,])])
table(idx<-rownames(fiveUTRhits87_mapq_filter) %in% unlist(tx87[rownames(d[1:6,])])$tx_name )

temp<-fiveUTRhits87_mapq_filter[idx,]

fiveUTRhits87_mapq_filter["Y39A1A.12",]
assays(fiveUTRhits87_mapq_filter["W03C9.4a",])$counts
assays(fiveUTRhits87_mapq_filter["W03C9.4c",])$counts
```



#old summarizeOverlaps section to be deleted
```{r summarizeOverlaps,eval=F}
load("ens87.rdata")
(fls <- list.files(".", pattern="*_q4.bam$",full=TRUE))
fls <- fls[!grepl("PE",fls)]
bamlst <- BamFileList(fls)
detectCores()
register(MulticoreParam(workers=detectCores()))
genehits82 <- summarizeOverlaps(ens82,bamlst,mode="Union",singleEnd=TRUE,ignore.strand=TRUE)
genehits82q4 <- summarizeOverlaps(ens82,bamlst,mode="Union",singleEnd=TRUE,ignore.strand=TRUE)
#genehits <- summarizeOverlaps(ens87,bamlst,mode="Union",singleEnd=TRUE,ignore.strand=TRUE)
head(assays(genehits82)$counts)
apply(assays(genehits82)$counts,2,sum)
save(genehits82,file= "genehits_ens82_WBcel235_010617.rdata")
save(genehits,file= "genehits_ens87_WBcel235_010617.rdata")
```

# Do some QC on read counts
```{r EDA, eval=T}

#load("genehits_ens82_WBcel235_111815.rdata")
#load("genehits_ens82_WBcel235_112515.rdata")
#load("genehits_ens87_WBcel235_010617.rdata")
#head(temp<-as.data.frame(assays(genehits)$counts))
head(temp<-as.data.frame(assays(genehits82)$counts))

summary(idx<-match(rownames(temp),type82$ensembl_gene_id))
temp$biotype<-type82[idx,"transcript_biotype"]
head(temp)

colnames(temp)<-sapply(strsplit(colnames(temp),"_"),function(x) x[1]) %>%
  gsub("2566","OMA-1_",.) %>% gsub("3923","LIN-41_",.) %>% gsub("IP4","OMA-1_042310",.)

colors<-c("#cbc9e2","#9e9ac8","#6a51a3",
          "#fdbe85","#fd8d3c","#d94701",
          "#56B4E9", "#009E73","#F0E442", "#0072B2",
          "black","red","green")

gather(temp,sample,counts,-biotype) %>%
  group_by(sample,biotype) %>%
  summarize(countSum=sum(counts)) %>%
  #filter(biotype!="rRNA") %>%
  #filter(biotype!="protein_coding") %>%
  #filter(biotype!="pseudogene") %>%
  ggplot(aes(x=sample,y=countSum,fill=biotype)) +
  scale_fill_manual(values=rev(colors),guide = guide_legend(reverse=TRUE)) +
  geom_bar(stat="identity") + theme_classic() + theme(axis.text.x = element_text(angle = 90,  hjust = 0)) 
```

## Build DEseq object excluding rRNA but w/o fitering for uniquely mapped reads
```{r}
load("genehits_ens82_WBcel235_010617.rdata")
apply(assays(genehits82)$counts,2,sum)
temp<-as.data.frame(log2((assays(genehits82)$counts)+1))
summary(temp)
rownames(temp[rowMeans(temp)>19,])

type<-getBM(filters="ensembl_gene_id",values=rownames(genehits82_mapq_filter), attributes=c("ensembl_gene_id","transcript_biotype"), mart=ensembl_82)
table(type$transcript_biotype)

idx<-match(rownames(genehits82),type$ensembl_gene_id)
head(idx)
genehits<-genehits82[type[idx,]$transcript_biotype != "rRNA",]
dim(genehits)
cds<-DESeqDataSet(genehits,design=~1)
#colnames(cds)<-sapply(strsplit(colnames(cds),"_"),function(x) x[1])
file_names<-colnames(cds)
colnames(cds)<-sapply(strsplit(colnames(cds),"_"),function(x) x[1]) %>%
  gsub("2566","OMA-1_",.) %>% gsub("3923","LIN-41_",.) %>% gsub("IP4","OMA-1_042310",.)

cds$tag<-as.factor(c("OMA-1","OMA-1","LysRZ","Lys","LIN-41","LIN-41","LysRZ","Lys","OMA-1_042310"))
cds$filename<-file_names
design(cds)<-(~tag)
colnames(cds)

#get normalization factors for cds
plotPCA( DESeqTransform( cds ),intgroup="tag")+
  ggtitle("Greenstein RNP Purification") + theme_bw()
cds<-DESeq(cds)
colData(cds)
round(colData(cds)$sizeFactor)


#subset samples we are working with
cds2<-cds[,c(1,2,3,5,6,7)]

plotPCA( DESeqTransform( cds2 ),intgroup="tag")+
  ggtitle("Greenstein RNP Purification") + theme_bw()

cds2$tag <- droplevels(cds2$tag)
cds2<-DESeq(cds2)
plotMA(cds2,ylim=c(-8,8))
resultsNames(cds)
```




"Mon Jan 30 14:36:09 2017"
## Build DESeq object with uniquely mapping reads that are not rRNA
```{r}
apply(assays(genehits82_mapq_filter)$counts,2,sum)
temp<-as.data.frame(log2((assays(genehits82_mapq_filter)$counts)+1))
summary(temp)
rownames(temp[rowMeans(temp)>19,])
#annotations[grep("WBGene00004622",annotations$ensembl_gene_id),]
#View(biomaRt::listAttributes(ensembl_82))

#getBM(filters="ensembl_gene_id",values=rownames(temp[rowMeans(temp)>19,]), attributes=c("ensembl_gene_id","transcript_biotype"), mart=ensembl_82)
type<-getBM(filters="ensembl_gene_id",values=rownames(genehits82_mapq_filter), attributes=c("ensembl_gene_id","transcript_biotype"), mart=ensembl_82)
table(type$transcript_biotype)

idx<-match(rownames(genehits82_mapq_filter),type$ensembl_gene_id)
head(idx)
genehits<-genehits82_mapq_filter[type[idx,]$transcript_biotype != "rRNA",]
dim(genehits)
cds<-DESeqDataSet(genehits,design=~1)
#colnames(cds)<-sapply(strsplit(colnames(cds),"_"),function(x) x[1])
file_names<-colnames(cds)
colnames(cds)<-sapply(strsplit(colnames(cds),"_"),function(x) x[1]) %>%
  gsub("2566","OMA-1_",.) %>% gsub("3923","LIN-41_",.) %>% gsub("IP4","OMA-1_042310",.)

cds$tag<-as.factor(c("OMA-1","OMA-1","LysRZ","Lys","LIN-41","LIN-41","LysRZ","Lys","OMA-1_042310"))
cds$filename<-file_names
design(cds)<-(~tag)
colnames(cds)

#get normalization factors for cds
plotPCA( DESeqTransform( cds ),intgroup="tag")+
  ggtitle("Greenstein RNP Purification") + theme_bw()
cds<-DESeq(cds)
colData(cds)
round(colData(cds)$sizeFactor)
x<-colData(cds)
save(x,file="/mnt/smb/micah/greenste_coldata.rdata")

#subset samples we are working with
cds2<-cds[,c(1,2,3,5,6,7)]

plotPCA( DESeqTransform( cds2 ),intgroup="tag")+
  ggtitle("Greenstein RNP Purification") + theme_bw()

cds2$tag <- droplevels(cds2$tag)
cds2<-DESeq(cds2)
plotMA(cds2,ylim=c(-8,8))
resultsNames(cds2)

res<-results(cds,lfcThreshold=2,contrast=c("tag","OMA-1","LysRZ"))
res<-as.data.frame(res)
res<-res[!is.na(res$padj),]
res<-res[abs(res$log2FoldChange) > 2,]
dim(res) #1190
head(res<-res[with(res,order(padj)),])
```

```{r}
attrib<-listAttributes(ensembl_82)
attrib[grep("Symbol",attrib[,2]),]
head(annotations<-getBM(filters="ensembl_gene_id",values=names(ens82), attributes=c("ensembl_gene_id","wikigene_name","transcript_biotype","external_gene_name"), mart=ensembl_82))

getBM(filters="ensembl_gene_id",values="WBGene00017498",attributes=c("ensembl_gene_id","wikigene_name","external_gene_name"),mart=ensembl_82)

idx<-match(rownames(res),annotations$ensembl_gene_id)



res$name<-annotations[idx,"wikigene_name"]
res$type<-annotations[idx,"transcript_biotype"]
head(res,20)


write.csv(res,file="RNP_purification_112018.csv",quote=F)
```

#get normalized counts for 042310 Oma-1 sample
```{r}
cds<-DESeq(cds)
cds_counts<-counts(cds,normalized=TRUE)
head(oma1_042310<-log2(cds_counts[,9]+1)-0.5*log2(cds_counts[,3]+1)-0.5*log2(cds_counts[,7]+1))
```


#This sections calculates the Oma-1 and LIN-41 logFCs
```{r}
res<-results(cds2,lfcThreshold=2,contrast=c("tag","OMA-1","LysRZ"))
res2<-results(cds2,lfcThreshold=2,contrast=c("tag","LIN-41","LysRZ"))
dim(res2)

head(logFCs<-cbind(oma1_042310,res$log2FoldChange,res2$log2FoldChange))
colnames(logFCs)<-c("Oma-1 042310","Oma-1 IP","Lin-41 IP")
plot(logFCs[,2],logFCs[,3],cex=0.5,pch=16,xlab="Oma-1",ylab="Lin-41",
     main="Enrichment Relative to Ribozero Lysate")
plot(logFCs[,2],logFCs[,1],cex=0.5,pch=16,xlab="Oma-1",ylab="Oma-1 042310",
     main="Enrichment Relative to Ribozero Lysate")

#remove the NAs
temp<-as.data.frame(logFCs[!(is.na(logFCs[,2]) | is.na(logFCs[,3])),])
dim(temp)
dim(temp<-temp[rowVars(temp)>6.5,])
idx<-match(rownames(temp),annotations$ensembl_gene_id)
temp$name<-annotations[idx,]$external_gene_name
pheatmap(temp[,1:3],labels_row=annotations[idx,]$external_gene_name)

temp<-as.data.frame(logFCs[!(is.na(logFCs[,2]) | is.na(logFCs[,3])),])
dim(temp<-temp[temp[,2]>4 | temp[,3] > 4,])
pheatmap(temp[,1:3],show_rownames=F)

#output
dim(temp<-as.data.frame(logFCs[!(is.na(logFCs[,2]) | is.na(logFCs[,3])),]))
idx<-match(rownames(temp),annotations$ensembl_gene_id)
temp$name<-annotations[idx,]$external_gene_name
#dim(temp<-temp[temp[,2]>4 | temp[,3] > 4,])
head(temp)
colnames(temp)<-c("a","b","c","d")
temp<-temp[with(temp,order(-c)),]
colnames(temp)<-c("Oma-1 Spike et al, 2014a","Oma-1 IP", "Lin-41 IP", "Gene Name")
head(temp)
dim(temp)
#write.csv(temp,file="enrichment_relative_to_LysateRZ_112715.csv",quote=F)
write.csv(temp,file="enrichment_relative_to_LysateRZ_013117.csv",quote=F)
```

##100216
```{r}
#temp<-read.csv("enrichments_fpkms_112715 - Original Enrichment Relative to LysateRZ.csv",stringsAsFactors = F)
#plot(temp$Oma.1.IP,temp$Lin.41.IP,cex=0.5,pch=16,xlab="Oma-1",ylab="Lin-41",
#     main="Enrichment Relative to Ribozero Lysate")
#smoothScatter(temp$Oma.1.IP,temp$Lin.41.IP,cex=0.5,pch=16,xlab="Oma-1",ylab="Lin-41",
#     main="Enrichment Relative to Ribozero Lysate")

plot(temp$`Oma-1 IP`,temp$`Lin-41 IP`,cex=0.5,pch=16,xlab="Oma-1",ylab="Lin-41",
     main="Enrichment Relative to Ribozero Lysate")
smoothScatter(temp$`Oma-1 IP`,temp$`Lin-41 IP`,cex=0.5,pch=16,xlab="Oma-1",ylab="Lin-41",
     main="Enrichment Relative to Ribozero Lysate")


Lab.palette <- colorRampPalette(c("white", "orange", "red"), space = "Lab")
#jpeg("oma1_vs_lin41_scatter.jpg",width=600,height=600,type="cairo")
#svglite("testplot.svg",width=8,height=8)
i.s<-smoothScatter(temp$`Oma-1 IP`,temp$`Lin-41 IP`, colramp = Lab.palette,
                     ## pch=NA: do not draw them
                     xlab=("Log2 Oma-1 Enrichment"),
                     ylab=("Log2 Lin-41 Enrichment"),
                     nrpoints = 100, ret.selection=TRUE)
#identify(temp$Oma.1.IP,temp$Lin.41.IP, labels=temp$Gene.Name)
gene_list<-c("spn-4","cdc-25.3","lin-29","zif-1","rnp-1")
for ( g in gene_list ){
  x<-temp[grep(g,temp$'Gene Name'),'Oma-1 IP']
  y<-temp[grep(g,temp$'Gene Name'),'Lin-41 IP']
  print(x)
  print(y)
  print(g)
  text(x,y,g)
  }
abline(h=2,col="red")
abline(v=2,col="red")
#dev.off()
abline(a=2,b=1,col="red")
abline(a=-2,b=1,col="red")
## label the 20 very lowest-density points,the "outliers" (with obs.number):
i.20 <- i.s[1:20]
text(x[i.20,], labels = i.20, cex= 0.75)


#meanSdPlot from kernSmooth::meanSdme

mydata <- data.frame(x=exp(rnorm(10000)), y=exp(rnorm(10000)))

ggplot(temp, aes(x=Oma.1.IP, y=Lin.41.IP)) +
   geom_point(size=0.5) +
   stat_density2d(aes(alpha=..density.., fill=..density..), geom="tile", contour=FALSE) + theme_bw()

ggplot(temp, aes(x=`Oma-1 IP`, y=`Lin-41 IP`)) + geom_point(size=0.5) +
  stat_density2d(geom="tile", aes(fill=..density..^0.15, alpha=1), contour=FALSE) +
  stat_density2d(geom="tile", aes(fill=..density..^0.15,  alpha=ifelse(..density..^0.15<0.4,0,1)), contour=FALSE) +
  scale_fill_gradientn(colours = colorRampPalette(c("white", blues9))(256)) + theme_bw()
```




"Mon Dec 14 13:55:35 2015"
```{r}
system("wget 'http://trace.ncbi.nlm.nih.gov/Traces/sra/sra.cgi?save=efetch&rettype=runinfo&db=sra&term=SRP041461' -O sra.csv")
system("wget 'http://trace.ncbi.nlm.nih.gov/Traces/sra/sra.cgi?save=efetch&rettype=runinfo&db=sra&term=SRX527966' -O sra.csv")
View(sra<-read.csv("sra.csv",stringsAsFactors=F))
sqlfile<-"/mnt/gcd/sradb/SRAmetadb.sqlite"
if(!file.exists(sqlfile)) sqlfile <<- getSRAdbFile()
sra_con <- dbConnect(SQLite(),sqlfile)
conversion <- sraConvert('SRP041461', sra_con = sra_con )
samples<-unique(conversion$experiment)
dbListFields(sra_con,"experiment")
sample_names<-list()
#dbGetQuery(sra_con,paste0("select title from experiment where experiment_accession = 'SRX527966'"))

for (i in 1:length(samples)) {
  x<- dbGetQuery(sra_con,paste0("select title from experiment where experiment_accession = '",samples[i],"'"))
  sample_names[[samples[i]]]<-sapply(strsplit(x[1,1],":|;"),function(x) x[2])
}

for (i in samples) {
  x<-conversion[grep(i,conversion$experiment),"run"]
  print(paste0("samtools merge ", i,".bam ", x[1],".bam ",x[2],".bam ",x[3],".bam ",x[4],".bam "))
}

dbGetQuery(sra_con,paste0("select title from experiment where experiment_accession = 'SRX527966'"))
```

```{r}
#Quantitate reads in all the merged SRX files (4 SRR files per SRX file)
(fls <- list.files("../kimble", pattern=glob2rx("SRX*.bam"),full=TRUE))
bamlst <- BamFileList(fls)
detectCores()
register(MulticoreParam(workers=detectCores()))
kimble82 <- summarizeOverlaps(ens82,bamlst,mode="Union",singleEnd=TRUE,ignore.strand=TRUE)
save(kimble82,file="kimble_ens82_WBcel235_011017.rdata")
load("./kimble_ens82_WBcel235_011017.rdata")
kimble<-DESeqDataSet(kimble82,design=~1)
file_names<-colnames(kimble)
colnames(kimble)<-gsub("\\.bam","",file_names)
kimble$title<-unlist(sample_names)[colnames(kimble)]
kimble$filename<-file_names
kimble$sex<-factor(substr(kimble$title,2,4))
#sperm [fem-3(q96)] or oocytes [fog-2(q71)]
design(kimble)<-(~sex)
colData(kimble)
plotPCA( DESeqTransform( kimble ),intgroup="sex")+
  ggtitle("Kimble Data") + theme_bw()
kimble<-DESeq(kimble)
plotMA(kimble)

res<-results(kimble,lfcThreshold=1,contrast=c("sex","q96","q71"))
res<-as.data.frame(res)
res<-res[!is.na(res$padj),]
#res<-res[abs(res$log2FoldChange) > 2,]
#dim(res) #1190
head(res<-res[with(res,order(-log2FoldChange)),])

idx<-match(rownames(res),annotations$ensembl_gene_id)
res$name<-annotations[idx,"wikigene_name"]
res$external_name<-annotations[idx,"external_gene_name"]
res$type<-annotations[idx,"transcript_biotype"]
head(res,20)

genes_of_interest<-c("cpb-1","fog-1","fog-3","oma-1","pie-1","him-3","spo-11","C38C6.8","mab-3","spn-4")
res[res$name %in% genes_of_interest,]

plot(res$log2FoldChange,-1*log10(res$padj),cex=0.5,pch=16)

#Get average make and female fpkms
kimble_fpkm<-as.data.frame(fpkm(kimble))
(oocytes<-which(colData(kimble)[colnames(kimble_fpkm),"sex"]=="q71"))
(sperm<-which(colData(kimble)[colnames(kimble_fpkm),"sex"]=="q96"))
kimble_fpkm$oocytes<-apply(kimble_fpkm,1,function(x) mean(x[oocytes]))
kimble_fpkm$sperm<-apply(kimble_fpkm,1,function(x) mean(x[sperm]))

#merge with res
idx<-match(rownames(res),rownames(kimble_fpkm))
res$oocyte_fpkm<-round(kimble_fpkm[idx,"oocytes"],3)
res$sperm_fpkm<-round(kimble_fpkm[idx,"sperm"],3)

head(kimble_data<-res)
```



Do a merge to upload
```{r}
head(temp<-read.csv(file="enrichment_relative_to_LysateRZ_112715.csv",stringsAsFactors=F,row.names=1))
dim(temp)
dim(kimble_data)
idx<-match(rownames(temp),rownames(kimble_data))
temp$gonad.baseMean<-kimble_data[idx,"baseMean"]
temp$gonadLog2FoldChange<-kimble_data[idx,"log2FoldChange"]
temp$gonadPadj<-kimble_data[idx,"padj"]
head(temp)
temp[temp$Gene.Name %in% genes_of_interest,]

write.csv(temp,file="enrichment_relative_to_LysateRZ_withGonadEnrichment_121815.csv",quote=F)

```

"Tue Jan 10 11:38:52 2017"
I have sat down to try to write the RNAseq section of the Results. I am trying to do as much as I can myself, to make this most efficient for you. I am beginning by doing what I can and organizing things and making an outline (before actually writing). I am looking through the data to see if I can distinguish the obvious lin-41 somatic targets in the heterochronic pathway (i.e., lin-29 and mab-10) from oocyte targets. I am wondering whether I could get a file very similar to the enrichments_fpkms_112715 file but with the  Kimble log2 oocyte enrichment in the spreadsheet with the fpkms in the ip and lysate. The reason is I am thinking of binning mRNAs enriched in the LIN-41 IP by abundance in the IP and then in each group scoring oocyte enrichment (like Figure 3C in Spike et al, 2014a). Does this make sense?

```{r}
#need abundance in the lin-41 IP samples vs enrichment 
temp2<-as.data.frame(temp)

head(f<-as.data.frame(fpkm(cds,robust=TRUE)))
(oma1_ip_columns<-which(grepl("OMA-1_IP",colnames(f))))
(lin41_ip_columns<-which(grepl("LIN-41_IP",colnames(f))))
f$oma1_ip_mean<-apply(f,1,function(x) mean(x[oma1_ip_columns]))
f$lin41_ip_mean<-apply(f,1,function(x) mean(x[lin41_ip_columns]))

#add kimble mean FPKM values to our data
summary(idx<-match(rownames(temp2),rownames(kimble_data)))
temp2$sperm_fpkm<-kimble_data[idx,"sperm_fpkm"]
temp2$oocyte_fpkm<-kimble_data[idx,"oocyte_fpkm"]
temp2$kimble_log2FC<-kimble_data[idx,"log2FoldChange"]
temp2$kimble_padj<-kimble_data[idx,"padj"]


idx<-match(rownames(temp2),rownames(f))
temp2$oma1_ip_mean<-f[idx,"oma1_ip_mean"]
temp2$lin41_ip_mean<-f[idx,"lin41_ip_mean"]
head(temp2)


plot(log2(temp2$oma1_ip_mean),log2(temp2$oocyte_fpkm),cex=0.5,pch=16)
plot(log2(temp2$oma1_ip_mean),temp2$kimble_log2FC,cex=0.5,pch=16)
abline(h=0,col="red")

plot(log2(temp2$lin41_ip_mean),log2(temp2$oocyte_fpkm),cex=0.5,pch=16)
plot(log2(temp2$lin41_ip_mean),temp2$kimble_log2FC,cex=0.5,pch=16)
abline(h=0,col="red")

temp2[1:12,]
g1<-temp2 %>%  
  mutate(kimble_log2FC=ifelse(is.na(kimble_log2FC),0,kimble_log2FC)) %>% # set NA's to 0 difference
  ggplot(aes(x=log2(lin41_ip_mean),y=`Lin-41 IP`)) +
  ggtitle("Lin-41 Enrichment vs Abundance") + xlab("Log2 (Lin-41 FPKM)") + ylab("Log2 Lin-41 Enrichment in IP")+
  xlim(c(-12,20)) + ylim(c(-10,10)) +
  geom_point(size=0.5,aes(colour = kimble_log2FC)) + 
  scale_colour_gradient2(low = "red2", mid = "white", high = "darkblue") +
  theme_bw()

g2<-temp2 %>%  
  mutate(kimble_log2FC=ifelse(is.na(kimble_log2FC),0,kimble_log2FC)) %>% # set NA's to 0 difference
  ggplot(aes(x=log2(oma1_ip_mean),y=`Oma-1 IP`)) +
  xlim(c(-12,20)) + ylim(c(-10,10)) +
  ggtitle("Oma-1 Enrichment vs Abundance") + xlab("Log2 (Oma-1 FPKM)") + ylab("Log2 Oma-1 Enrichment in IP")+
  geom_point(size=0.5,aes(colour = kimble_log2FC)) + 
  scale_colour_gradient2(low = "red2", mid = "white", high = "darkblue") +
  theme_bw()
svglite("testplot.svg",width=10,height=8)
grid.arrange(g2,g1,ncol=2)
dev.off()
```

Make a barplot showing the bias for Oocyte specific genes in the IP'd RNPs.  The point here is to determine if the RNPs are loaded in a biased way with transcripts involved in oogenesis.  Spike et al originally showed this by seeing how many oogenic genes were in each "bin" of the transcripts present in the IP.  But very low abundance transcripts that are oocyte enriched may have large enrichments in the RNPs that would be lost by just binning the IP transcripts by abundance.  Probably need to look at it both ways and see if a clear trend is present.

Rethink - these RNPs are clearly going to be loaded with oocyte genes since the expression of the tagged construct is mostly oocyte.   It is really just a control to show that the enriched genes are (IP vs lysate) are 
```{r}
#try a volcano on the kimble data
temp2 %>%  
 # mutate(kimble_log2FC=ifelse(is.na(kimble_log2FC),0,kimble_log2FC)) %>% # set NA's to 0 difference
  ggplot(aes(x=kimble_log2FC,y=(-1)*log10(kimble_padj))) +
  ggtitle("Lin-41 Enrichment vs Abundance") + xlab("Log2 (Lin-41 FPKM)") + ylab("Log2 Lin-41 Enrichment in IP")+
#  xlim(c(-12,20)) + ylim(c(-10,10)) +
  geom_point(size=0.5,aes(colour = `Oma-1 IP`)) + 
  scale_colour_gradient2(low = "red2", mid = "white", high = "darkblue") +
  theme_bw()
```

```{r barplot}
#make the barplot
temp2 %>% 
  dplyr::select(oma1_ip_mean,lin41_ip_mean,oocyte_fpkm) %>% 
  mutate(oocyte_fpkm=ifelse(is.na(oocyte_fpkm),0.001,oocyte_fpkm)) %>% 
  gather(gene,fpkm,oma1_ip_mean,lin41_ip_mean) %>% 
  mutate(high_oocyte = ifelse(oocyte_fpkm > 1,"Oogenic","not Oogenic")) %>% 
  mutate(bin=cut(log2(fpkm+0.001),5,labels=c("e","d","c","b","a"))) %>% 
  mutate(bin=factor(bin,levels=c("a","b","c","d","e"))) %>% 
  group_by(gene,bin) %>% 
  ggplot(aes(x=gene,fill=high_oocyte))+
  geom_bar(stat="count") +facet_grid(.~bin) + theme_bw() +
  xlab("High Abundance  ... Low Abundance") +
  ylab("Number of Transcripts, color by Oocyte Expression Level")  +
  theme(axis.text.x = element_text(angle = 90,  hjust = 0)) 

```

## FPKM Table
Add annotations and export the FPKMs
```{r}
#add annotations to f
idx<-match(rownames(f),annotations$ensembl_gene_id)
f$name<-annotations[idx,]$external_gene_name
head(f)
f[grep("mab-10",f$name),]
f[grep("pud-4",f$name),]
write.csv(f,file="fpkm_112715.csv",quote=F)
```

#Venn Diagrams
```{r}
lfc<-2
#x<-nrow(temp[temp$Lin.41.IP > lfc,])
#y<-nrow(temp[temp$Oma.1.IP > lfc,])
#z<-nrow(temp[temp$Lin.41.IP > lfc & temp$Oma.1.IP > lfc,])
(x<-nrow(temp[temp$`Lin-41 IP`> lfc,]))
(y<-nrow(temp[temp$`Oma-1 IP` > lfc,]))
(z<-nrow(temp[temp$`Lin-41 IP` > lfc & temp$`Oma-1 IP` > lfc,]))
#svglite("testplot.svg")
grid.newpage()
vennplot <- draw.pairwise.venn(x,y,z, c("Lin-41", "Oma-1"))
#dev.off()

write.csv(temp[temp$Lin.41.IP > lfc | temp$Oma.1.IP > lfc,],file="enrichment_4fold.csv",quote=F)

temp[temp$Gene.Name %in% c("cdc-25.3", "zif-1","rnp-1"),]


```


"Fri Jan 27 13:31:58 2017"
## GO Analysis
```{r eval=F}
downloader::download("ftp://ftp.wormbase.org/pub/wormbase/releases/WS257/ONTOLOGY/gene_association.WS257.wb",
                     "gene_association.WS257.wb")
downloader::download("ftp://ftp.wormbase.org/pub/wormbase/releases/WS257/ONTOLOGY/gene_ontology.WS257.obo",
                     "gene_ontology.WS257.obo")

ga<-readr::read_delim("gene_association.WS257.wb",skip=3,delim="\t",col_names=F)

#Scan in OBO file to get Biological Process terms
t1<-readLines("gene_ontology.WS257.obo",n=-1L)
terms<-list()

for (i in seq_along(1:length(t1))) {
  if (t1[i]=="[Term]") {
    gt<-substr(t1[i+1],5,14)
    name<-gsub("name: ","",t1[i+2])
    if (t1[i+3]=="namespace: biological_process") { 
      #print (paste0(gt,":  ",name)) 
      terms[[gt]]<-name
    }
    }
}
go_terms<-unlist(terms)

head(ga)
wb_go<-ga[ga$'X5' %in% names(go_terms),c('X2','X5')]
colnames(wb_go)<-c("ensembl","GO ID")
dim(wb_go<-as.data.frame(wb_go))
#remove duplicated entries (possibly from multiple lines of evidence)
dim(wb_go<-wb_go[!duplicated(wb_go),])

wb_go.list<-split(wb_go$`GO ID`,wb_go$ensembl)

#List Terms for a gene
annotations[grep("lin-41",annotations$wikigene_name),]
go_terms[wb_go.list[["WBGene00003026"]]]

annotations[grep("lin-29",annotations$wikigene_name),]
go_terms[wb_go.list[["WBGene00003015"]]]

#List Genes for a Term
annotations[annotations$ensembl_gene_id %in% wb_go[grep("GO:0045138",wb_go$`GO ID`),"ensembl"],]

save(wb_go,go_terms,file="wb_257_goterms.rdata")
```

#Run goseq
```{r}
load("wb_257_goterms.rdata")
length(expressed_genes<-rownames(res[res$baseMean > 1 & !is.na(res$log2FoldChange) & !is.na(res$padj),]))
dim(wb_go<-wb_go[wb_go$ensembl %in% expressed_genes,])
wb_go.list<-split(wb_go$`GO ID`,wb_go$ensembl)


#bias data
bd<-sum(width(reduce(ens82)))
bd<-bd[names(bd) %in% expressed_genes]
length(bd)
length(expressed_genes)
head(bd)

#degs
temp<-res[expressed_genes,]
temp_deg<-subset(temp,temp$padj < 0.05 & abs(temp$log2FoldChange ) > 1)
degs<-as.numeric(rownames(temp) %in% rownames(temp_deg))
names(degs)<-rownames(temp)
length(degs)
table(degs)

listGO<-function(goid) {
  print(go_terms[goid])
  tg<-wb_go[grep(goid,wb_go$`GO ID`),"ensembl",drop=F]
  idx<-match(tg$ensembl,annotations$ensembl_gene_id)
  tg$symbol<-annotations[idx,"wikigene_name"]
  tg$deg<-degs[tg$ensembl]
  as.data.frame(tg)
}

listGO("GO:0045138")

#GOseq
pwf<-nullp(degs,bias.data=bd)

GO.wall<-goseq(pwf,gene2cat=wb_go.list)

GO.wall[1:20,c("category","term","numDEInCat","numInCat","over_represented_pvalue")]

head(GO.wall,20) %>%
  dplyr::mutate(term=factor(term,levels=rev(term))) %>%
ggplot(aes(x=term,y=-log10(over_represented_pvalue))) +
  geom_bar(stat="identity",fill="red") +
  coord_flip() + xlab("") +
  theme_bw() 

```



