---
title: "Lin41 Analysis"
author: "Micah Gearhart"
date: "1/6/2017"
output:
  html_document:
    keep_md: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r libraries,eval=T,results='hide',message=FALSE,warning=FALSE}
library("biomaRt")
library("GenomicFeatures")
library("Rsamtools")

library("GenomicFeatures")
library("Rsamtools")
library("GenomicAlignments")
library("BiocParallel")
library("DESeq2")
library("genefilter")
library("pheatmap")
library("SRAdb")
library("goseq")


library("ggplot2")
library("dplyr")
library("tidyr")
library("magrittr")
library("RColorBrewer")
library("gridExtra")
library("motifStack")
library("VennDiagram")

(ts<-format(Sys.time(), "%a_%b_%d_%Y_%H%M"))
cbPalette <- c("#999999", "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7")

cbPalette2<-c("#cbc9e2","#9e9ac8","#6a51a3", "#fdbe85","#fd8d3c","#d94701",
          "#56B4E9", "#009E73","#F0E442", "#0072B2", "black","red","green")

wls<- function(x) { paste0(seqnames(x),":",start(x),"-",end(x))}
```

## Download Annotation files for Ensembl 87
```{r biomaRt, eval=F}

ensembl_87 = useMart(biomart="ENSEMBL_MART_ENSEMBL", host="useast.ensembl.org", path="/biomart/martservice",dataset="celegans_gene_ensembl")
txdb<-makeTxDbFromBiomart(biomart="ENSEMBL_MART_ENSEMBL",host="useast.ensembl.org",dataset="celegans_gene_ensembl")

#tx87<-transcriptsBy(txdb, "gene")
ens87<-exonsBy(txdb,by="gene")

#annotations
anno87<-getBM(filters="ensembl_gene_id",values=names(ens87), attributes=c("ensembl_gene_id","wikigene_name","transcript_biotype","external_gene_name"), mart=ensembl_87)

save(ens87,anno87,file="ens87.rdata")
```


## Count 2x for Ensembl 87
```{r summarizeOverlaps, eval=F}
load("ens87.rdata")
mapq_filter <- function(features, reads, ignore.strand = FALSE, inter.feature=TRUE)
{
    require(GenomicAlignments) # needed for parallel evaluation
    Union(features, reads[mcols(reads)$mapq >= 4],
          ignore.strand, inter.feature)
}

param <- ScanBamParam(what="mapq")

#select single end bam files that are not filtered by samtools (q4)
(fls <- list.files(".", pattern=glob2rx("*.bam"),full=TRUE))
(fls<-fls[!grepl("q4",fls)])
(fls<-fls[!grepl("PE",fls)])
(fls<-fls[!grepl("IP4_4_23_10.bam",fls)])
bamlst <- BamFileList(fls)

#summarizeOverlaps with multimappers
genehits87<-summarizeOverlaps(ens87,bamlst,mode="Union",singleEnd=TRUE,ignore.strand=TRUE)
apply(assays(genehits87)$counts,2,sum)

#with multimapping Filter
genehits87_mapq_filter<-summarizeOverlaps(ens87,bamlst,mode="mapq_filter", param=param,
                                 singleEnd=TRUE,ignore.strand=TRUE)
apply(assays(genehits87_mapq_filter)$counts,2,sum)

#just check to make sure this is the same as samtools filtering
#(fls <- list.files(".", pattern=glob2rx("*_q4.bam"),full=TRUE))
#bamlst <- BamFileList(fls)
#genehits82q4<-summarizeOverlaps(ens82,bamlst,mode="Union",singleEnd=TRUE,ignore.strand=TRUE)
#apply(assays(genehits82q4)$counts,2,sum)
#all.equal(unname(apply(assays(genehits82_mapq_filter)$counts,2,sum)),unname(apply(assays(genehits82q4)$counts,2,sum)))

save(genehits87,genehits87_mapq_filter,file=paste0("genehits87_",ts,".rdata"))
```

## Compare the biotype Distributions across samples with and without a MAPQ filter
```{r eda,eval=T,fig.height=10,fig.width=6}
load("ens87.rdata")
load("genehits87_Mon_Feb_06_2017_1446.rdata")

counts87<-as.data.frame(assays(genehits87)$counts)
counts87q4<-as.data.frame(assays(genehits87_mapq_filter)$counts)
                          
summary(idx<-match(rownames(counts87),anno87$ensembl_gene_id))
counts87$biotype<-anno87[idx,"transcript_biotype"]
counts87q4$biotype<-anno87[idx,"transcript_biotype"]

g1<-tidyr::gather(counts87,sample,counts,-biotype) %>%
  dplyr::group_by(sample,biotype) %>%
  dplyr::summarize(countSum=sum(counts)) %>%
  ggplot(aes(x=sample,y=countSum,fill=biotype)) + ggtitle("Counts with Ensembl87") +
  scale_fill_manual(values=rev(cbPalette2),guide = guide_legend(reverse=TRUE)) +
  geom_bar(stat="identity") + theme_classic() + theme(axis.text.x = element_text(angle = 90,  hjust = 0)) 

g2<-tidyr::gather(counts87q4,sample,counts,-biotype) %>%
  dplyr::group_by(sample,biotype) %>%
  dplyr::summarize(countSum=sum(counts)) %>%
  ggplot(aes(x=sample,y=countSum,fill=biotype)) + ggtitle("Counts with Ensembl87 with Mapping Quality Filter") +
  scale_fill_manual(values=rev(cbPalette2),guide = guide_legend(reverse=TRUE)) +
  geom_bar(stat="identity") + theme_classic() + theme(axis.text.x = element_text(angle = 90,  hjust = 0)) 

g1
g2
#grid.arrange(g1,g2,ncol=1)
```


## Filter out rRNA genes and analyze for Enrichment relative to Input
```{r PCA,eval=T}
summary(idx<-match(rownames(genehits87),anno87$ensembl_gene_id))
dim(dds87<-DESeqDataSet(genehits87[anno87[idx,]$transcript_biotype != "rRNA",],design=~1))
dim(dds87q4<-DESeqDataSet(genehits87_mapq_filter[anno87[idx,]$transcript_biotype != "rRNA",],design=~1))

file_names<-colnames(dds87)
colnames(dds87)<-sapply(strsplit(colnames(dds87),"_"),function(x) x[1]) %>%
  gsub("2566","OMA-1_",.) %>% gsub("3923","LIN-41_",.) %>% gsub("IP4","OMA-1_042310",.)

colnames(dds87q4)<-sapply(strsplit(colnames(dds87q4),"_"),function(x) x[1]) %>%
  gsub("2566","OMA-1_",.) %>% gsub("3923","LIN-41_",.) %>% gsub("IP4","OMA-1_042310",.)

dds87$tag<-as.factor(c("OMA-1","OMA-1","LysRZ","Lys","LIN-41","LIN-41","LysRZ","Lys","OMA-1_042310"))
dds87q4$tag<-as.factor(c("OMA-1","OMA-1","LysRZ","Lys","LIN-41","LIN-41","LysRZ","Lys","OMA-1_042310"))

design(dds87)<-(~tag)
design(dds87q4)<-(~tag)

plotPCA( DESeqTransform( dds87 ),intgroup="tag")+ ggtitle("PCA without MapQ Filter") + theme_bw()
plotPCA( DESeqTransform( dds87q4 ),intgroup="tag")+ ggtitle("PCA with MapQ Filter") + theme_bw()

```

## Run Differential Expression on both dds objects
```{r DESeq2,eval=T}
dds87<-DESeq(dds87)
plotMA(dds87,ylim=c(-8,8))
resultsNames(dds87)

res_Oma1_dds87<-results(dds87,lfcThreshold=2,contrast=c("tag","OMA-1","LysRZ"))
res_Lin41_dds87<-results(dds87,lfcThreshold=2,contrast=c("tag","LIN-41","LysRZ"))

#Estimate LogFC for single replicate Oma-1 sample
dds87_counts<-counts(dds87,normalized=TRUE)
res_oma1_042310_dds87<-log2(dds87_counts[,9]+1)-0.5*log2(dds87_counts[,3]+1)-0.5*log2(dds87_counts[,7]+1)

res_dds87<-cbind(res_oma1_042310_dds87,res_Oma1_dds87$log2FoldChange,res_Lin41_dds87$log2FoldChange)
colnames(res_dds87)<-c("Oma-1 042310","Oma-1 IP","Lin-41 IP")

dds87q4<-DESeq(dds87q4)
plotMA(dds87q4,ylim=c(-8,8))
resultsNames(dds87q4)

res_Oma1_dds87q4<-results(dds87q4,lfcThreshold=2,contrast=c("tag","OMA-1","LysRZ"))
res_Lin41_dds87q4<-results(dds87q4,lfcThreshold=2,contrast=c("tag","LIN-41","LysRZ"))

#Estimate LogFC for single replicate Oma-1 sample
dds87q4_counts<-counts(dds87q4,normalized=TRUE)
res_oma1_042310_dds87q4<-log2(dds87q4_counts[,9]+1)-0.5*log2(dds87q4_counts[,3]+1)-0.5*log2(dds87q4_counts[,7]+1)

res_dds87q4<-as.data.frame(cbind(res_oma1_042310_dds87q4,res_Oma1_dds87q4$log2FoldChange,res_Lin41_dds87q4$log2FoldChange))
colnames(res_dds87q4)<-c("Oma-1 042310","Oma-1 IP","Lin-41 IP")
```

## Quick look at enrichment in old vs new data
```{r quickscatter,eval=T}
idx<-!(is.na(res_dds87q4[,2]) | is.na(res_dds87q4[,3]))
corr_oma_v_lin<-cor(res_dds87q4[idx,2],res_dds87q4[idx,3],method="spearman")
corr_oma_v_oma<-cor(res_dds87q4[idx,2],res_dds87q4[idx,1],method="spearman")

plot(res_dds87q4[,2],res_dds87q4[,3],cex=0.5,pch=16,xlab="Oma-1",ylab="Lin-41",
     main="Enrichment Relative to Ribozero Lysate")
text(-6,5,paste0("Spearman correlation:  ",round(corr_oma_v_lin,3)),col="blue")


plot(res_dds87q4[,2],res_dds87q4[,1],cex=0.5,pch=16,xlab="Oma-1",ylab="Oma-1 042310",
     main="Enrichment Relative to Ribozero Lysate")
text(-6,5,paste0("Spearman correlation:  ",round(corr_oma_v_oma,3)),col="blue")
```



## MAKE A Master Table with MAPQ FLAG, Biotype, and mean FPKM per sample
```{r masterTable,eval=T}
#Add annotation information

idx<-match(rownames(res_dds87q4),anno87$ensembl_gene_id)
res_dds87q4$biotype<-anno87[idx,"transcript_biotype"]
res_dds87q4$name<-anno87[idx,"wikigene_name"]
res_dds87q4$external_name<-anno87[idx,"external_gene_name"]


lfc<-2  # also note not using abs() here -- just want enrichment
p<-0.05
length(oma1_genes<-rownames(subset(res_Oma1_dds87,log2FoldChange > lfc & padj < p)))
length(oma1q4_genes<-rownames(subset(res_Oma1_dds87q4,log2FoldChange > lfc & padj < p)))

length(lin41_genes<-rownames(subset(res_Lin41_dds87,log2FoldChange > lfc & padj < p)))
length(lin41q4_genes<-rownames(subset(res_Lin41_dds87q4,log2FoldChange > lfc & padj < p)))

#Find genes with different result due to MAPQ filter

length(mapq_effected_genes<-setdiff(sort(unique(c(oma1_genes,lin41_genes))),sort(unique(c(oma1q4_genes,lin41q4_genes)))))
res_dds87q4$mapq_effect <- rownames(res_dds87q4) %in% mapq_effected_genes

#Add FPKMs
dds87q4_fpkm<-as.data.frame(fpkm(dds87q4))
(lysRZ_columns<-which(colData(dds87q4)[colnames(dds87q4),"tag"]=="LysRZ"))
(oma1_columns<-which(colData(dds87q4)[colnames(dds87q4),"tag"]=="OMA-1"))
(lin41_columns<-which(colData(dds87q4)[colnames(dds87q4),"tag"]=="LIN-41"))
dds87q4_fpkm$lysRZ<-apply(dds87q4_fpkm,1,function(x) mean(x[lysRZ_columns]))
dds87q4_fpkm$oma1<-apply(dds87q4_fpkm,1,function(x) mean(x[oma1_columns]))
dds87q4_fpkm$lin41<-apply(dds87q4_fpkm,1,function(x) mean(x[lin41_columns]))

idx<-match(rownames(res_dds87q4),rownames(dds87q4_fpkm))
res_dds87q4$oma1_042310_fpkm<-round(dds87q4_fpkm[idx,"OMA-1_042310"],2)
res_dds87q4$lysRZ_fpkm<-round(dds87q4_fpkm[idx,"lysRZ"],2)
res_dds87q4$oma1_fpkm<-round(dds87q4_fpkm[idx,"oma1"],2)
res_dds87q4$lin41_fpkm<-round(dds87q4_fpkm[idx,"lin41"],2)

```

## Heatmaps of High Log2FC or High Variance Genes
```{r heatmap,eval=T,fig.height=10,fig.width=6}
#log2 fold change in eiter sample
dim(temp<-res_dds87q4[!(is.na(res_dds87q4[,2]) | is.na(res_dds87q4[,3])),])
dim(temp<-temp[temp[,2]>2 | temp[,3] > 2,])
pheatmap(temp[,1:3],show_rownames=F)

#just samples with highest variance between Oma-1 and Lin-41
dim(temp<-temp[rowVars(temp[,2:3])>8.5,]) #arbitrary cutoff to get a readable gene list
pheatmap(temp[,1:3],labels_row=temp$external_name)
```

#Venn Diagrams - with p-values
```{r vennDiagrams,eval=T}

x1<-length(oma1_genes)
y1<-length(lin41_genes)
z1<-sum(oma1_genes %in% lin41_genes)
grid.newpage()
vennplot <- draw.pairwise.venn(x1,y1,z1, c("Oma-1", "Lin-41"))

x2<-length(oma1q4_genes)
y2<-length(lin41q4_genes)
z2<-length(overlapq4<-oma1q4_genes[oma1q4_genes %in% lin41q4_genes])
x21<-length(oma1q4_genes_exclusive<-oma1q4_genes[!oma1q4_genes %in% overlapq4])
y21<-length(lin41q4_genes_exclusive<-lin41q4_genes[!lin41q4_genes %in% overlapq4])
grid.newpage()
vennplot <- draw.pairwise.venn(x2,y2,z2, c("Oma-1", "Lin-41"))
```

## Use scatterplot to visualize relationships in Venn Diagram
```{r scatterplot,eval=T}
plot(res_dds87q4[,2],res_dds87q4[,3],cex=0.5,pch=16,ylim=c(-7,7),xlim=c(-7,7),main="Oma-1 vs Lin-41 Enrichment", 
     ylab="LIN-41 Log2Fold Change",xlab="OMA-1 Log2Fold Change",
     col=ifelse(rownames(res_dds87q4) %in% overlapq4,"blue",
                ifelse(rownames(res_dds87q4) %in% oma1q4_genes_exclusive,"green",
                ifelse(rownames(res_dds87q4) %in% lin41q4_genes_exclusive,"red","black"))))
abline(a=0,b=1,col="red")
```

## Visualise as smoothScatter

```{r smoothScatter,eval=T}
Lab.palette <- colorRampPalette(c("white", "orange", "red"), space = "Lab")

#jpeg("oma1_vs_lin41_scatter.jpg",width=600,height=600,type="cairo")
#svglite("testplot.svg",width=8,height=8)
i.s<-smoothScatter(res_dds87q4$`Oma-1 IP`,res_dds87q4$`Lin-41 IP`, colramp = Lab.palette,
                     ## pch=NA: do not draw them
                     xlab=("Log2 Oma-1 Enrichment"),
                     ylab=("Log2 Lin-41 Enrichment"),
                     nrpoints = 100, ret.selection=TRUE,main="Relative Enrichment in IP's")
#identify(temp$Oma.1.IP,temp$Lin.41.IP, labels=temp$Gene.Name)
gene_list<-c("spn-4","cdc-25.3","lin-29","zif-1","rnp-1")
for ( g in gene_list ){
  x<-res_dds87q4[grep(g,res_dds87q4$'name'),'Oma-1 IP']
  y<-res_dds87q4[grep(g,res_dds87q4$'name'),'Lin-41 IP']
  print(x)
  print(y)
  print(g)
  text(x,y,g)
  }
abline(h=2,col="red")
abline(v=2,col="red")
```


#Now run GO Analysis on each section of Venn Diagram
## GO Analysis
```{r download_GO, eval=F}
downloader::download("ftp://ftp.wormbase.org/pub/wormbase/releases/WS257/ONTOLOGY/gene_association.WS257.wb",
                     "gene_association.WS257.wb")
downloader::download("ftp://ftp.wormbase.org/pub/wormbase/releases/WS257/ONTOLOGY/gene_ontology.WS257.obo",
                     "gene_ontology.WS257.obo")

ga<-readr::read_delim("gene_association.WS257.wb",skip=3,delim="\t",col_names=F)

#Scan in OBO file to get Biological Process terms
t1<-readLines("gene_ontology.WS257.obo",n=-1L)
terms<-list()

for (i in seq_along(1:length(t1))) {
  if (t1[i]=="[Term]") {
    gt<-substr(t1[i+1],5,14)
    name<-gsub("name: ","",t1[i+2])
    if (t1[i+3]=="namespace: biological_process") { 
      #print (paste0(gt,":  ",name)) 
      terms[[gt]]<-name
    }
    }
}
go_terms<-unlist(terms)

head(ga)
wb_go<-ga[ga$'X5' %in% names(go_terms),c('X2','X5')]
colnames(wb_go)<-c("ensembl","GO ID")
dim(wb_go<-as.data.frame(wb_go))
#remove duplicated entries (possibly from multiple lines of evidence)
dim(wb_go<-wb_go[!duplicated(wb_go),])

wb_go.list<-split(wb_go$`GO ID`,wb_go$ensembl)

#List Terms for a gene
anno87[grep("lin-41",anno87$wikigene_name),]
go_terms[wb_go.list[["WBGene00003026"]]]

anno87[grep("lin-29",anno87$wikigene_name),]
go_terms[wb_go.list[["WBGene00003015"]]]

#List Genes for a Term
anno87[anno87$ensembl_gene_id %in% wb_go[grep("GO:0045138",wb_go$`GO ID`),"ensembl"],]

save(wb_go,go_terms,file="wb_257_goterms.rdata")
```

## GO Analysis of each VENN category
```{r GO,eval=T}
load("wb_257_goterms.rdata")
length(expressed_genes<-rownames(res_Lin41_dds87q4[res_Lin41_dds87q4$baseMean > 1,]))

#alternatively insist on enough data to define p-value
#length(expressed_genes<-rownames(res_Lin41_dds87q4[res_Lin41_dds87q4$baseMean > 1 & !is.na(res_Lin41_dds87q4$log2FoldChange) & !is.na(res_Lin41_dds87q4$padj) & !is.na(res_Oma1_dds87q4$log2FoldChange) & !is.na(res_Oma1_dds87q4$padj),]))

dim(wb_go<-wb_go[wb_go$ensembl %in% expressed_genes,])
wb_go.list<-split(wb_go$`GO ID`,wb_go$ensembl)

listGO<-function(goid,degs) {
  print(go_terms[goid])
  tg<-wb_go[grep(goid,wb_go$`GO ID`),"ensembl",drop=F]
  idx<-match(tg$ensembl,anno87$ensembl_gene_id)
  tg$symbol<-anno87[idx,"wikigene_name"]
  tg$deg<-degs[tg$ensembl]
  as.data.frame(tg)
}

#degs_temp<-rep(NA,length(expressed_genes))
#names(degs_temp)<-expressed_genes
#listGO("GO:0045138",degs_temp)

#bias data
bd<-sum(width(reduce(ens87)))
bd<-bd[names(bd) %in% expressed_genes]

#GOseq - Lin41 exclusive
table(degs_lin41<-as.numeric(expressed_genes %in% lin41q4_genes_exclusive ))
names(degs_lin41)<-expressed_genes
lin41_GO<-goseq(nullp(degs_lin41,bias.data=bd),gene2cat=wb_go.list)
lin41_GO[1:10,c("category","term","numDEInCat","numInCat","over_represented_pvalue")]

listGO("GO:1903864",degs_lin41)

#GOseq - Oma1 exclusive
table(degs_oma1<-as.numeric(expressed_genes %in% oma1q4_genes_exclusive ))
names(degs_oma1)<-expressed_genes
oma1_GO<-goseq(nullp(degs_oma1,bias.data=bd),gene2cat=wb_go.list)
oma1_GO[1:10,c("category","term","numDEInCat","numInCat","over_represented_pvalue")]

listGO("GO:0007219",degs_oma1)

#GOseq - overlap
table(degs_overlap<-as.numeric(expressed_genes %in% overlapq4 ))
names(degs_overlap)<-expressed_genes
overlap_GO<-goseq(nullp(degs_overlap,bias.data=bd),gene2cat=wb_go.list)
overlap_GO[1:10,c("category","term","numDEInCat","numInCat","over_represented_pvalue")]

#horizontal bar plot
head(lin41_GO,10) %>%
  dplyr::mutate(term=factor(term,levels=rev(term))) %>%
  ggplot(aes(x=term,y=-log10(over_represented_pvalue))) +
  geom_bar(stat="identity",fill="red") +
  coord_flip() + xlab("") + ggtitle("Enriched GO terms LIN-41 Exclusive Set") +
  theme_bw()

listGO("GO:0007538",degs_overlap)

```


## MOTIF ANALYSIS

# Download UTR sequences from biomaRt
```{r utr_biomart,eval=F}
#get transcript identifiers for all expressed genes
expressed_transcripts<-getBM(filters="ensembl_gene_id",values=expressed_genes, attributes=c("ensembl_gene_id","ensembl_transcript_id"), mart=ensembl_87)

seq3utr<-getSequence(id = expressed_transcripts$ensembl_transcript_id, type = "ensembl_transcript_id", seqType = "3utr", mart = ensembl_87)
seq5utr<-getSequence(id = expressed_transcripts$ensembl_transcript_id, type = "ensembl_transcript_id", seqType = "5utr", mart = ensembl_87)
save(seq3utr,seq5utr,expressed_transcripts,file="utr_sequences.rdata")
```

# Search for OMA-1 and LIN-41 Motifs in 3' UTRs
```{r threeUTR,eval=T}
load("utr_sequences.rdata")

#clean up UTR sequences
seq3utr<-seq3utr[seq3utr$`3utr`!="Sequence unavailable",]
seq3utr<-seq3utr[nchar(seq3utr$`3utr`) >=8,]  # get rid of really short 3'UTRs

# add geneid to dataframe
summary(idx<-match(seq3utr$ensembl_transcript_id,expressed_transcripts$ensembl_transcript_id))
seq3utr$ensembl_gene_id<-expressed_transcripts[idx,"ensembl_gene_id"]
sum(duplicated(paste0(seq3utr$`3utr`,seq3utr$ensembl_transcript_id)))
sum(duplicated(paste0(seq3utr$`3utr`,seq3utr$ensembl_gene_id)))
seq3utr<-seq3utr[!duplicated(paste0(seq3utr$`3utr`,seq3utr$ensembl_gene_id)),]

seq3utrBS<-DNAStringSet(seq3utr$`3utr`)
names(seq3utrBS)<-seq3utr$ensembl_transcript_id
summary(width(seq3utrBS))
seq3utr<-seq3utr[,-1]  #drop sequence from dataframe

#Count TAW motif and RNACompete Motif in each transcript
seq3utr$taw<-vcountPattern("TAW",seq3utrBS,fixed=FALSE)
table(as.logical(seq3utr$taw))

#Count RNAcompete motif in each transcript
rnaCompeteMotif<-RNAStringSet(c("GCAAAGC","GUAAAAC","GUAUAAC","GCAUAGC","CUUUAAG"))
temp<-consensusMatrix(rnaCompeteMotif)[1:4,]
plotMotifLogo(new("pfm",mat=t(t(temp[1:4,])*1/colSums(temp[1:4,])), name="RNAcompeteMotif"))

lin41_consensusMatrix<-consensusMatrix(DNAStringSet(rnaCompeteMotif))[1:4,]

seq3utr$loedige<-unlist(lapply(seq3utrBS,function(x) countPWM(lin41_consensusMatrix,x,min.score="90%")))
table(as.logical(seq3utr$loedige))

#oma-1 TAW
temp<-seq3utr %>% group_by(ensembl_gene_id) %>% 
  summarize(taw_log=as.logical(sum(taw)),loedige_log=as.logical(sum(loedige))) %>% as.data.frame

a<-sum(temp[temp$ensembl_gene_id %in% oma1q4_genes,"taw_log"])
b<-sum(temp[!temp$ensembl_gene_id %in% oma1q4_genes,"taw_log"])
c<-sum(oma1q4_genes %in% temp$ensembl_gene_id)
d<-sum(expressed_genes %in% temp$ensembl_gene_id)

(x<-matrix(c(d-b-c,b,c-a,a),ncol=2))
(oma1_taw_3p_test<-fisher.test(x))

#oma-1 loedige
a<-sum(temp[temp$ensembl_gene_id %in% oma1q4_genes,"loedige_log"])
b<-sum(temp[!temp$ensembl_gene_id %in% oma1q4_genes,"loedige_log"])
c<-sum(oma1q4_genes %in% temp$ensembl_gene_id)
d<-sum(expressed_genes %in% temp$ensembl_gene_id)

(x<-matrix(c(d-b-c,b,c-a,a),ncol=2))
(oma1_loedige_3p_test<-fisher.test(x))

#lin41 TAW
a<-sum(temp[temp$ensembl_gene_id %in% lin41q4_genes,"taw_log"])
b<-sum(temp[!temp$ensembl_gene_id %in% lin41q4_genes,"taw_log"])
c<-sum(lin41q4_genes %in% temp$ensembl_gene_id)
d<-sum(expressed_genes %in% temp$ensembl_gene_id)

(x<-matrix(c(d-b-c,b,c-a,a),ncol=2))
(lin41_taw_3p_test<-fisher.test(x))

#lin41 loedige
a<-sum(temp[temp$ensembl_gene_id %in% lin41q4_genes,"loedige_log"])
b<-sum(temp[!temp$ensembl_gene_id %in% lin41q4_genes,"loedige_log"])
c<-sum(lin41q4_genes %in% temp$ensembl_gene_id)
d<-sum(expressed_genes %in% temp$ensembl_gene_id)

(x<-matrix(c(d-b-c,b,c-a,a),ncol=2))
(lin41_loedige_3p_test<-fisher.test(x))

x<-round(data.frame(pvalue=c(oma1_taw_3p_test$p.value,oma1_loedige_3p_test$p.value,lin41_taw_3p_test$p.value,lin41_loedige_3p_test$p.value)),5)
rownames(x)<-c("TAW motif enrichment in 3' UTR of Oma1 Enriched Genes",
               "RNAcompete Lin-41 motif enrichment in 3' UTR of Oma1 Enriched Genes",
               "TAW motif enrichment in 3' UTR of Lin-41 Enriched Genes",
               "RNACompete Lin-41 motif enrichment in 3' UTR of Lin-41 Enriched Gene")
knitr::kable(x)
```

## Search for OMA-1 and LIN-41 Motifs in 5' UTRs 
```{r fiveUTR,eval=T}
#clean up UTR sequences
seq5utr<-seq5utr[seq5utr$`5utr`!="Sequence unavailable",]
seq5utr<-seq5utr[nchar(seq5utr$`5utr`) >=8,]  # get rid of really short 3'UTRs

# add geneid to dataframe
summary(idx<-match(seq5utr$ensembl_transcript_id,expressed_transcripts$ensembl_transcript_id))
seq5utr$ensembl_gene_id<-expressed_transcripts[idx,"ensembl_gene_id"]
sum(duplicated(paste0(seq5utr$`5utr`,seq5utr$ensembl_transcript_id)))
sum(duplicated(paste0(seq5utr$`5utr`,seq5utr$ensembl_gene_id)))
seq5utr<-seq5utr[!duplicated(paste0(seq5utr$`5utr`,seq5utr$ensembl_gene_id)),]

seq5utrBS<-DNAStringSet(seq5utr$`5utr`)
names(seq5utrBS)<-seq5utr$ensembl_transcript_id
summary(width(seq5utrBS))
seq5utr<-seq5utr[,-1]  #drop sequence from dataframe

#Count TAW motif and RNACompete Motif in each transcript
seq5utr$taw<-vcountPattern("TAW",seq5utrBS,fixed=FALSE)
table(as.logical(seq5utr$taw))

#Count RNAcompete motif in each transcript
seq5utr$loedige<-unlist(lapply(seq5utrBS,function(x) countPWM(lin41_consensusMatrix,x,min.score="90%")))
table(as.logical(seq5utr$loedige))

#oma-1 TAW
temp<-seq5utr %>% group_by(ensembl_gene_id) %>% 
  summarize(taw_log=as.logical(sum(taw)),loedige_log=as.logical(sum(loedige))) %>% as.data.frame

a<-sum(temp[temp$ensembl_gene_id %in% oma1q4_genes,"taw_log"])
b<-sum(temp[!temp$ensembl_gene_id %in% oma1q4_genes,"taw_log"])
c<-sum(oma1q4_genes %in% temp$ensembl_gene_id)
d<-sum(expressed_genes %in% temp$ensembl_gene_id)

(x<-matrix(c(d-b-c,b,c-a,a),ncol=2))
(oma1_taw_5p_test<-fisher.test(x))

#oma-1 loedige
a<-sum(temp[temp$ensembl_gene_id %in% oma1q4_genes,"loedige_log"])
b<-sum(temp[!temp$ensembl_gene_id %in% oma1q4_genes,"loedige_log"])
c<-sum(oma1q4_genes %in% temp$ensembl_gene_id)
d<-sum(expressed_genes %in% temp$ensembl_gene_id)

(x<-matrix(c(d-b-c,b,c-a,a),ncol=2))
(oma1_loedige_5p_test<-fisher.test(x))

#lin41 TAW
a<-sum(temp[temp$ensembl_gene_id %in% lin41q4_genes,"taw_log"])
b<-sum(temp[!temp$ensembl_gene_id %in% lin41q4_genes,"taw_log"])
c<-sum(lin41q4_genes %in% temp$ensembl_gene_id)
d<-sum(expressed_genes %in% temp$ensembl_gene_id)

(x<-matrix(c(d-b-c,b,c-a,a),ncol=2))
(lin41_taw_5p_test<-fisher.test(x))

#lin41 loedige
a<-sum(temp[temp$ensembl_gene_id %in% lin41q4_genes,"loedige_log"])
b<-sum(temp[!temp$ensembl_gene_id %in% lin41q4_genes,"loedige_log"])
c<-sum(lin41q4_genes %in% temp$ensembl_gene_id)
d<-sum(expressed_genes %in% temp$ensembl_gene_id)

(x<-matrix(c(d-b-c,b,c-a,a),ncol=2))
(lin41_loedige_5p_test<-fisher.test(x))

x<-round(data.frame(pvalue=c(oma1_taw_5p_test$p.value,oma1_loedige_5p_test$p.value,lin41_taw_5p_test$p.value,lin41_loedige_5p_test$p.value)),5)
rownames(x)<-c("TAW motif enrichment in 5' UTR of Oma1 Enriched Genes",
               "RNAcompete Lin-41 motif enrichment in 5' UTR of Oma1 Enriched Genes",
               "TAW motif enrichment in 5' UTR of Lin-41 Enriched Genes",
               "RNACompete Lin-41 motif enrichment in 5' UTR of Lin-41 Enriched Gene")
knitr::kable(x)
```


## MEME de novo explore UTR's of LIN-41 exclusive set

```{r run_meme,eval=F}

length(lin41q4_exclusive_3p_BS<-seq3utrBS[seq3utr[seq3utr$ensembl_gene_id %in% lin41q4_genes_exclusive,"ensembl_transcript_id"]])
length(lin41_3p_BS<-seq3utrBS[seq3utr[seq3utr$ensembl_gene_id %in% lin41q4_genes,"ensembl_transcript_id"]])

motifSet_lin41_3utr_exclusive <- runMEME(lin41q4_exclusive_3p_BS, binary="/usr/ngs/bin/meme/bin/meme",arguments=list("-nmotifs"=3,"-maxsize"=1e16))
motifSet_lin41_3utr_full <- runMEME(lin41_3p_BS, binary="/usr/ngs/bin/meme/bin/meme",arguments=list("-nmotifs"=3,"-maxsize"=1e16,,"-w"=7))
save(motifSet_lin41_3utr,motifSet_lin41_3utr_full,file="motifsets_lin41_3pUTR.rdata")
```


## Analyze Meme findings
```{r analyze_meme,eval=F}
load("motifsets_lin41_3pUTR.rdata")

motif2pfm<-function(m) {
  temp<-consensusMatrix(m)
  #rownames(temp)<-c("A","C","G","U")
  new("pfm",mat=t(t(temp[1:4,])*1/colSums(temp[1:4,])), name="3'UTR motif")
  }

plotMotifLogo(motif2pfm(motifSet_lin41_3utr_exclusive[[1]]))

seq5utrBS<-seq5utrBS[width(seq5utrBS)>= 8]
names(seq5utrBS)<-paste0("seq5utr_",1:length(seq5utrBS))
writeXStringSet(seq5utrBS,file="test5p.fasta")
motifSet_5utr <- runMEME(seq5utrBS, binary="/usr/ngs/bin/meme/bin/meme",arguments=list("-nmotifs"=3,"-maxsize"=10000000000000000,"-w"=5))
motifSet_5utr <- runMEME(seq5utrBS, binary="/usr/ngs/bin/meme/bin/meme",arguments=list("-nmotifs"=3,"-maxsize"=10000000000000000))
consensusMatrix(motifSet_5utr)

(temp<-consensusMatrix(motifSet_5utr)[[3]])
rownames(temp)<-c("A","C","G","U")
plotMotifLogo(new("pfm",mat=t(t(temp[1:4,])*1/colSums(temp[1:4,])), name="3'UTR motif"))

```

# KIMBLE DATA

"Mon Dec 14 13:55:35 2015"
```{r kimble_geoQuery,eval=F}
system("wget 'http://trace.ncbi.nlm.nih.gov/Traces/sra/sra.cgi?save=efetch&rettype=runinfo&db=sra&term=SRP041461' -O sra.csv")
system("wget 'http://trace.ncbi.nlm.nih.gov/Traces/sra/sra.cgi?save=efetch&rettype=runinfo&db=sra&term=SRX527966' -O sra.csv")
View(sra<-read.csv("sra.csv",stringsAsFactors=F))
sqlfile<-"/mnt/gcd/sradb/SRAmetadb.sqlite"
if(!file.exists(sqlfile)) sqlfile <<- getSRAdbFile()
sra_con <- dbConnect(SQLite(),sqlfile)
conversion <- sraConvert('SRP041461', sra_con = sra_con )
samples<-unique(conversion$experiment)
dbListFields(sra_con,"experiment")
sample_names<-list()
#dbGetQuery(sra_con,paste0("select title from experiment where experiment_accession = 'SRX527966'"))

for (i in 1:length(samples)) {
  x<- dbGetQuery(sra_con,paste0("select title from experiment where experiment_accession = '",samples[i],"'"))
  sample_names[[samples[i]]]<-sapply(strsplit(x[1,1],":|;"),function(x) x[2])
}

for (i in samples) {
  x<-conversion[grep(i,conversion$experiment),"run"]
  print(paste0("samtools merge ", i,".bam ", x[1],".bam ",x[2],".bam ",x[3],".bam ",x[4],".bam "))
}

dbGetQuery(sra_con,paste0("select title from experiment where experiment_accession = 'SRX527966'"))
save(sample_names,file="kimble_sample_names.rdata")
```

```{r summarizeOverlaps_kimble,eval=F}
#Quantitate reads in all the merged SRX files (4 SRR files per SRX file)
(fls <- list.files("../kimble", pattern=glob2rx("SRX*.bam"),full=TRUE))
bamlst <- BamFileList(fls,yieldSize=1e5)
detectCores()
register(MulticoreParam(workers=detectCores()))

kimble87 <-summarizeOverlaps(ens87,bamlst,mode="Union", singleEnd=TRUE,ignore.strand=TRUE)
apply(assays(kimble87)$counts,2,sum)

kimble87_mapq_filter<-summarizeOverlaps(ens87,bamlst,mode="mapq_filter", param=param,
                                 singleEnd=TRUE,ignore.strand=TRUE)
apply(assays(kimble87_mapq_filter)$counts,2,sum)

save(kimble87,kimble87_mapq_filter,file=paste0("kimble87_WBcel235_",ts,".rdata"))
```

```{r analyzeKimbleData,eval=T}
load("kimble87_WBcel235_Mon_Feb_06_2017_2213.rdata")
load("kimble_sample_names.rdata")
kimble<-DESeqDataSet(kimble87_mapq_filter,design=~1)
file_names<-colnames(kimble)
colnames(kimble)<-gsub("\\.bam","",file_names)
kimble$title<-unlist(sample_names)[colnames(kimble)]
kimble$filename<-file_names
kimble$sex<-factor(substr(kimble$title,2,4))
#sperm [fem-3(q96)] or oocytes [fog-2(q71)]
design(kimble)<-(~sex)
colData(kimble)
plotPCA( DESeqTransform( kimble ),intgroup="sex")+
  ggtitle("Kimble Data") + theme_bw()
kimble<-DESeq(kimble)
plotMA(kimble,ylim=c(-14,14))

res_kimble<-results(kimble,lfcThreshold=1,contrast=c("sex","q96","q71"))
res_kimble<-as.data.frame(res_kimble)
res_kimble<-res_kimble[!is.na(res_kimble$padj),]
#res<-res[abs(res$log2FoldChange) > 2,]
#dim(res) #1190
head(res_kimble<-res_kimble[with(res_kimble,order(-log2FoldChange)),])

idx<-match(rownames(res_kimble),anno87$ensembl_gene_id)
res_kimble$name<-anno87[idx,"wikigene_name"]
res_kimble$external_name<-anno87[idx,"external_gene_name"]
res_kimble$type<-anno87[idx,"transcript_biotype"]


plot(res_kimble$log2FoldChange,-1*log10(res_kimble$padj),cex=0.5,pch=16)

#Get average make and female fpkms
kimble_fpkm<-as.data.frame(fpkm(kimble))
(oocytes<-which(colData(kimble)[colnames(kimble_fpkm),"sex"]=="q71"))
(sperm<-which(colData(kimble)[colnames(kimble_fpkm),"sex"]=="q96"))
kimble_fpkm$oocytes<-apply(kimble_fpkm,1,function(x) mean(x[oocytes]))
kimble_fpkm$sperm<-apply(kimble_fpkm,1,function(x) mean(x[sperm]))

#merge with res
idx<-match(rownames(res_kimble),rownames(kimble_fpkm))
res_kimble$oocyte_fpkm<-round(kimble_fpkm[idx,"oocytes"],3)
res_kimble$sperm_fpkm<-round(kimble_fpkm[idx,"sperm"],3)

genes_of_interest<-c("cpb-1","fog-1","fog-3","oma-1","pie-1","him-3","spo-11","C38C6.8","mab-3","spn-4")
res_kimble[res_kimble$name %in% genes_of_interest,]
```

## Merge Kimble Data with Master Table and Export

```{r exportFPKM,eval=T}
idx<-match(rownames(res_dds87q4),rownames(res_kimble))
res_dds87q4$gonad.baseMean<-res_kimble[idx,"baseMean"]
res_dds87q4$gonadLog2FoldChange<-res_kimble[idx,"log2FoldChange"]
res_dds87q4$gonadPadj<-res_kimble[idx,"padj"]
res_dds87q4$oocyte_fpkm<-res_kimble[idx,"oocyte_fpkm"]
res_dds87q4$sperm_fpkm<-res_kimble[idx,"sperm_fpkm"]
head(res_dds87q4)
res_dds87q4[res_dds87q4$name %in% genes_of_interest,]

write.csv(temp,file=paste0("enrichment_relative_to_LysateRZ_withGonadEnrichment_",ts,".csv"),quote=F)

```

## Oocyte Expression levels and RNP enrichment

```{r oocyte_abundance,eval=T}

plot(log2(res_dds87q4$oma1_fpkm),log2(res_dds87q4$oocyte_fpkm),cex=0.5,pch=16)
plot(log2(res_dds87q4$oma1_fpkm),res_dds87q4$gonadLog2FoldChange,cex=0.5,pch=16)
abline(h=0,col="red")

plot(log2(res_dds87q4$lin41_fpkm),log2(res_dds87q4$oocyte_fpkm),cex=0.5,pch=16)
plot(log2(res_dds87q4$lin41_fpkm),res_dds87q4$gonadLog2FoldChange,cex=0.5,pch=16)
abline(h=0,col="red")

g3<-res_dds87q4 %>%  
  mutate(gonadLog2FoldChange=ifelse(is.na(gonadLog2FoldChange),0,gonadLog2FoldChange)) %>% # set NA's to 0 difference
  ggplot(aes(x=log2(lin41_fpkm),y=`Lin-41 IP`)) +
  ggtitle("Lin-41 Enrichment vs Abundance") + xlab("Log2 (Lin-41 FPKM)") + ylab("Log2 Lin-41 Enrichment in IP")+
  xlim(c(-12,20)) + ylim(c(-10,10)) +
  geom_point(size=0.5,aes(colour = gonadLog2FoldChange)) + 
  scale_colour_gradient2(low = "red2", mid = "white", high = "darkblue") +
  theme_bw()

g4<-res_dds87q4 %>%  
  mutate(gonadLog2FoldChange=ifelse(is.na(gonadLog2FoldChange),0,gonadLog2FoldChange)) %>% # set NA's to 0 difference
  ggplot(aes(x=log2(oma1_fpkm),y=`Oma-1 IP`)) +
  xlim(c(-12,20)) + ylim(c(-10,10)) +
  ggtitle("Oma-1 Enrichment vs Abundance") + xlab("Log2 (Oma-1 FPKM)") + ylab("Log2 Oma-1 Enrichment in IP")+
  geom_point(size=0.5,aes(colour = gonadLog2FoldChange)) + 
  scale_colour_gradient2(low = "red2", mid = "white", high = "darkblue") +
  theme_bw()

g3
g4

#svglite("testplot.svg",width=10,height=8)
#grid.arrange(g4,g3,ncol=2)
#dev.off()
```

Make a barplot showing the bias for Oocyte specific genes in the IP'd RNPs.  The point here is to determine if the RNPs are loaded in a biased way with transcripts involved in oogenesis.  Spike et al originally showed this by seeing how many oogenic genes were in each "bin" of the transcripts present in the IP.  But very low abundance transcripts that are oocyte enriched may have large enrichments in the RNPs that would be lost by just binning the IP transcripts by abundance.  Probably need to look at it both ways and see if a clear trend is present.

## Barplot to show high Abundance Transcripts are more likely to be Oogenic
```{r barplot,eval=T}
#make the barplot
res_dds87q4 %>% 
  dplyr::select(oma1_fpkm,lin41_fpkm,oocyte_fpkm) %>% 
  mutate(oocyte_fpkm=ifelse(is.na(oocyte_fpkm),0.001,oocyte_fpkm)) %>% 
  gather(gene,fpkm,oma1_fpkm,lin41_fpkm) %>% 
  mutate(high_oocyte = ifelse(oocyte_fpkm > 1,"Oogenic","not Oogenic")) %>% 
  mutate(bin=cut(log2(fpkm+0.001),5,labels=c("e","d","c","b","a"))) %>% 
  mutate(bin=factor(bin,levels=c("a","b","c","d","e"))) %>% 
  group_by(gene,bin) %>% 
  ggplot(aes(x=gene,fill=high_oocyte))+
  geom_bar(stat="count") +facet_grid(.~bin) + theme_bw() +
  xlab("High Abundance  ... Low Abundance") +
  ylab("Number of Transcripts, color by Oocyte Expression Level")  +
  theme(axis.text.x = element_text(angle = 90,  hjust = 0)) 

```

Rethink - these RNPs are clearly going to be loaded with oocyte genes since the expression of the tagged construct is mostly oocyte.   It is really just a control to show that the enriched genes (IP vs lysate) are from the oocytes.

```{r oocyte_volcano,eval=T}
#try a volcano on the kimble data
g5<-res_dds87q4  %>%  
  mutate(oma1_log2FC=ifelse(is.na(`Oma-1 IP`),0,`Oma-1 IP`)) %>% # set NA's to 0 difference
  ggplot(aes(x=gonadLog2FoldChange,y=(-1)*log10(gonadPadj))) +
  ggtitle("Gonad Expression Changes") + xlab("Log2 Gonad Fold Change") + ylab("-log10(gonad p-value)")+
  geom_point(size=0.5,aes(colour = oma1_log2FC)) + 
  scale_colour_gradient2(low = "red2", mid = "white", high = "darkblue") +
  theme_bw()

g6<-res_dds87q4  %>%  
  mutate(lin41_log2FC=ifelse(is.na(`Oma-1 IP`),0,`Lin-41 IP`)) %>% # set NA's to 0 difference
  ggplot(aes(x=gonadLog2FoldChange,y=(-1)*log10(gonadPadj))) +
  ggtitle("Gonad Expression Changes") + xlab("Log2 Gonad Fold Change") + ylab("-log10(gonad p-value)")+
  geom_point(size=0.5,aes(colour = lin41_log2FC)) + 
  scale_colour_gradient2(low = "red2", mid = "white", high = "darkblue") +
  theme_bw()

g5
g6
#grid.arrange(g5,g6,ncol=2)
```



